# Six Degrees of Kevin Bacon
**Introduction** - Six Degrees of Kevin Bacon is a game based on the "six degrees of separation"
concept, which posits that any two people on Earth are six or fewer acquaintance links apart. Movie
buffs challenge each other to find the shortest path between an arbitrary actor and prolific actor
Kevin Bacon. It rests on the assumption that anyone involved in the film industry can be linked
through their film roles to Bacon within six steps.
The analysis of social networks can be a computationally intensive task, especially when dealing with
large volumes of data. It is also a challenging problem to devise a correct methodology to infer an
informative social network structure. Here, we will analyze a social network of actors and actresses
that co-participated in movies. We will do some simple descriptive analysis, and in the end try to
relate an actor/actress’s position in the social network with the success of the movies in which they
participate.

#### Rules & Notes - Please take your time to read the following points:

1. The submission deadline will be set for the 5th of June at 23:59.
2. It is acceptable that you **discuss** with your colleagues different approaches to solve each step of the problem set, but the assignment is individual. That is, you are responsible for writing your own code, and analysing the results. Clear cases of cheating will be penalized with 0 points in this assignment;
3. After review of your submission files, and before a mark is attributed, you might be called to orally defend your submission;
4. You will be scored first and foremost by the number of correct answers, secondly by the logic used in the trying to approach each step of the problem set;
5. Consider skipping questions that you are stuck in, and get back to them later;
6. Expect computations to take a few minutes to finish in some of the steps.
7. **IMPORTANT** It is expected you have developed skills beyond writting SQL queries. Any question where you directly write a SQL query (then for example create a temporary table and use spark.sql to pass the query) will receive a 25% penalty. Using the spark syntax (for example dataframe.select("\*").where("conditions")) is acceptable and does not incur this penalty.
8. **Questions** – Any questions about this assignment should be posted in the Forum@Moodle. The last class will be an open office session for anyone with questions concerning the assignment. 
9. **Delivery** - To fulfil this activity you will have to upload the following materials to Moodle:
    1. An exported IPython notebook. The notebook should be solved (have results displayed), but should contain all neccesary code so that when the notebook is run in databricks it should also replicate these results. This means the all data downloading and processing should be done in this notebook. It is also important you clearly indicate where your final answer to each question is when you are using multiple cells (for example you print "my final anwser is" before your answer or use cell comments).
    2. **Delivery** - You will also need to provide a signed statement of authorship, which is present in the last page;
    3. It is recommended you read the whole assignment before starting.
    4. You can add as many cells as you like to answer the questions.
    5. You can make use of caching or persisitng your RDDs or Dataframes, this may speed up performance.
    6. If you have trouble with graphframes in databricks (specifically the import statement) you need to make sure the graphframes package is installed on the cluster you are running. If you click home on the left, then click on the graphframes library which you loaded in Lab 9 you can install the package on your cluster (check the graphframes checkbox and click install)

#### Data Sources and Description
We will use data from IMDB. You can download raw datafiles
from https://datasets.imdbws.com. Note that the files are tab delimited (.tsv) You can find a
description of the each datafile in https://www.imdb.com/interfaces/

## Questions
### Data loading and preperation
Review the file descriptions and load the necessary data onto your databricks cluser and into spark dataframes. You will need to use shell commands to download the data, unzip the data, load the data into spark. Note that the data might require parsing and preprocessing to be ready for the questions below.

**Hints** You can use 'gunzip' to unzip the .tz files. The data files will then be tab seperated (.tsv), which you can load into a dataframe using the tab seperated option instead of the comma seperated option we have typically used in class: `.option(“sep”,”\t”)`

In [0]:
# Name: Francisco João Mourato Calha  |  id: m20210673

In [0]:
#libraries used

from graphframes import *
from pyspark.sql.types import *
from pyspark.sql import Row
from pyspark.sql.functions import isnull, when, count, col
from pyspark.ml import Pipeline
from pyspark.ml.feature import MinMaxScaler
import pyspark.ml.feature as MF
from pyspark.ml.regression import LinearRegression
from pyspark.ml.tuning import ParamGridBuilder
from pyspark.ml.tuning import CrossValidator
from pyspark.ml.feature import RobustScaler
from pyspark.ml.regression import GBTRegressor
from pyspark.ml.regression import RandomForestRegressor
from pyspark.ml.evaluation import RegressionEvaluator


In [0]:
#getting all the data and unzipping it

In [0]:
%sh

wget 'https://datasets.imdbws.com/name.basics.tsv.gz'
gunzip name.basics.tsv.gz

--2022-06-08 10:15:35--  https://datasets.imdbws.com/name.basics.tsv.gz
Resolving datasets.imdbws.com (datasets.imdbws.com)... 99.84.66.78, 99.84.66.98, 99.84.66.41, ...
Connecting to datasets.imdbws.com (datasets.imdbws.com)|99.84.66.78|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 228668577 (218M) [binary/octet-stream]
Saving to: ‘name.basics.tsv.gz’

     0K .......... .......... .......... .......... ..........  0% 7.79M 28s
    50K .......... .......... .......... .......... ..........  0% 5.51M 34s
   100K .......... .......... .......... .......... ..........  0% 8.79M 31s
   150K .......... .......... .......... .......... ..........  0% 31.2M 25s
   200K .......... .......... .......... .......... ..........  0% 13.1M 23s
   250K .......... .......... .......... .......... ..........  0% 45.6M 20s
   300K .......... .......... .......... .......... ..........  0% 13.4M 20s
   350K .......... .......... .......... .......... ..........  0% 16.7M 19s


In [0]:
%sh

wget 'https://datasets.imdbws.com/title.akas.tsv.gz'
gunzip title.akas.tsv.gz

--2022-06-08 10:15:57--  https://datasets.imdbws.com/title.akas.tsv.gz
Resolving datasets.imdbws.com (datasets.imdbws.com)... 99.84.66.56, 99.84.66.78, 99.84.66.98, ...
Connecting to datasets.imdbws.com (datasets.imdbws.com)|99.84.66.56|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 273528175 (261M) [binary/octet-stream]
Saving to: ‘title.akas.tsv.gz’

     0K .......... .......... .......... .......... ..........  0% 2.72M 96s
    50K .......... .......... .......... .......... ..........  0% 7.78M 65s
   100K .......... .......... .......... .......... ..........  0% 7.42M 55s
   150K .......... .......... .......... .......... ..........  0% 2.93M 63s
   200K .......... .......... .......... .......... ..........  0% 24.7M 53s
   250K .......... .......... .......... .......... ..........  0% 79.0M 45s
   300K .......... .......... .......... .......... ..........  0%  150M 38s
   350K .......... .......... .......... .......... ..........  0% 25.0M 35s
  

In [0]:
%sh

wget 'https://datasets.imdbws.com/title.basics.tsv.gz'
gunzip title.basics.tsv.gz

--2022-06-08 10:16:22--  https://datasets.imdbws.com/title.basics.tsv.gz
Resolving datasets.imdbws.com (datasets.imdbws.com)... 99.84.66.41, 99.84.66.56, 99.84.66.78, ...
Connecting to datasets.imdbws.com (datasets.imdbws.com)|99.84.66.41|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 156966383 (150M) [binary/octet-stream]
Saving to: ‘title.basics.tsv.gz’

     0K .......... .......... .......... .......... ..........  0% 13.9M 11s
    50K .......... .......... .......... .......... ..........  0% 4.56M 22s
   100K .......... .......... .......... .......... ..........  0% 13.7M 18s
   150K .......... .......... .......... .......... ..........  0% 6.79M 19s
   200K .......... .......... .......... .......... ..........  0% 12.9M 18s
   250K .......... .......... .......... .......... ..........  0% 13.2M 17s
   300K .......... .......... .......... .......... ..........  0% 21.7M 15s
   350K .......... .......... .......... .......... ..........  0% 12.6M 15

In [0]:
'''
%sh

wget 'https://datasets.imdbws.com/title.crew.tsv.gz'
gunzip title.crew.tsv.gz'''

Out[6]: "\n%sh\n\nwget 'https://datasets.imdbws.com/title.crew.tsv.gz'\ngunzip title.crew.tsv.gz"

In [0]:
'''
%sh

wget 'https://datasets.imdbws.com/title.episode.tsv.gz'
gunzip title.episode.tsv.gz'''

Out[7]: "\n%sh\n\nwget 'https://datasets.imdbws.com/title.episode.tsv.gz'\ngunzip title.episode.tsv.gz"

In [0]:
%sh

wget 'https://datasets.imdbws.com/title.principals.tsv.gz'
gunzip title.principals.tsv.gz

--2022-06-08 10:16:37--  https://datasets.imdbws.com/title.principals.tsv.gz
Resolving datasets.imdbws.com (datasets.imdbws.com)... 99.84.66.78, 99.84.66.98, 99.84.66.41, ...
Connecting to datasets.imdbws.com (datasets.imdbws.com)|99.84.66.78|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 400651552 (382M) [binary/octet-stream]
Saving to: ‘title.principals.tsv.gz’

     0K .......... .......... .......... .......... ..........  0% 3.94M 97s
    50K .......... .......... .......... .......... ..........  0% 8.23M 72s
   100K .......... .......... .......... .......... ..........  0% 9.01M 62s
   150K .......... .......... .......... .......... ..........  0% 25.1M 50s
   200K .......... .......... .......... .......... ..........  0% 21.5M 44s
   250K .......... .......... .......... .......... ..........  0% 24.7M 39s
   300K .......... .......... .......... .......... ..........  0% 19.2M 36s
   350K .......... .......... .......... .......... ..........  0% 

In [0]:
%sh

wget 'https://datasets.imdbws.com/title.ratings.tsv.gz'
gunzip title.ratings.tsv.gz

--2022-06-08 10:17:25--  https://datasets.imdbws.com/title.ratings.tsv.gz
Resolving datasets.imdbws.com (datasets.imdbws.com)... 99.84.66.56, 99.84.66.78, 99.84.66.98, ...
Connecting to datasets.imdbws.com (datasets.imdbws.com)|99.84.66.56|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 6244725 (6.0M) [binary/octet-stream]
Saving to: ‘title.ratings.tsv.gz’

     0K .......... .......... .......... .......... ..........  0% 7.54M 1s
    50K .......... .......... .......... .......... ..........  1% 8.51M 1s
   100K .......... .......... .......... .......... ..........  2% 8.74M 1s
   150K .......... .......... .......... .......... ..........  3% 28.1M 1s
   200K .......... .......... .......... .......... ..........  4% 28.3M 0s
   250K .......... .......... .......... .......... ..........  4% 23.2M 0s
   300K .......... .......... .......... .......... ..........  5% 47.6M 0s
   350K .......... .......... .......... .......... ..........  6% 28.0M 0s
   400

In [0]:
#we can see below that we already have the data needed

In [0]:
%sh ls

conf
eventlogs
ganglia
logs
metastore_db
name.basics.tsv
preload_class.lst
title.akas.tsv
title.basics.tsv
title.principals.tsv
title.ratings.tsv


In [0]:
name_basics = spark.read.format("csv") \
                        .option("header", "true") \
                        .option("inferSchema", "true") \
                        .option("delimiter", "\t") \
                        .load("file:/databricks/driver/name.basics.tsv")
name_basics.cache()

Out[12]: DataFrame[nconst: string, primaryName: string, birthYear: string, deathYear: string, primaryProfession: string, knownForTitles: string]

In [0]:
display(name_basics)

nconst primaryName birthYear deathYear primaryProfession knownForTitles nm0000001 Fred Astaire 1899 1987 soundtrack,actor,miscellaneous tt0053137,tt0031983,tt0050419,tt0072308 nm0000002 Lauren Bacall 1924 2014 actress,soundtrack tt0038355,tt0071877,tt0117057,tt0037382 nm0000003 Brigitte Bardot 1934 \N actress,soundtrack,music_department tt0054452,tt0057345,tt0049189,tt0056404 nm0000004 John Belushi 1949 1982 actor,soundtrack,writer tt0078723,tt0077975,tt0072562,tt0080455 nm0000005 Ingmar Bergman 1918 2007 writer,director,actor tt0050976,tt0050986,tt0060827,tt0083922 nm0000006 Ingrid Bergman 1915 1982 actress,soundtrack,producer tt0036855,tt0038109,tt0077711,tt0034583 nm0000007 Humphrey Bogart 1899 1957 actor,soundtrack,producer tt0034583,tt0037382,tt0042593,tt0043265 nm0000008 Marlon Brando 1924 2004 actor,soundtrack,director tt0070849,tt0068646,tt0047296,tt0078788 nm0000009 Richard Burton 1925 1984 actor,soundtrack,producer tt0057877,tt0059749,tt0061184,tt0087803 nm0000010 James Cagney 1899 1986 actor,soundtrack,director tt0029870,tt0042041,tt0035575,tt0031867 nm0000011 Gary Cooper 1901 1961 actor,soundtrack,producer tt0035896,tt0027996,tt0034167,tt0044706 nm0000012 Bette Davis 1908 1989 actress,soundtrack,make_up_department tt0056687,tt0042192,tt0035140,tt0031210 nm0000013 Doris Day 1922 2019 soundtrack,actress,producer tt0045591,tt0053172,tt0048317,tt0049470 nm0000014 Olivia de Havilland 1916 2020 actress,soundtrack tt0031381,tt0041452,tt0029843,tt0040806 nm0000015 James Dean 1931 1955 actor,miscellaneous tt0039123,tt0049261,tt0048545,tt0048028 nm0000016 Georges Delerue 1925 1992 composer,soundtrack,music_department tt0069946,tt8847712,tt0096320,tt0091763 nm0000017 Marlene Dietrich 1901 1992 soundtrack,actress,music_department tt0055031,tt0052311,tt0051201,tt0021156 nm0000018 Kirk Douglas 1916 2020 actor,producer,soundtrack tt0080736,tt0054331,tt0049456,tt0043338 nm0000019 Federico Fellini 1920 1993 writer,director,assistant_director tt0053779,tt0050783,tt0071129,tt0056801 nm0000020 Henry Fonda 1905 1982 actor,producer,soundtrack tt0050083,tt0032551,tt0082846,tt0051207 nm0000021 Joan Fontaine 1917 2013 actress,soundtrack,producer tt0040536,tt0034248,tt0032976,tt0035751 nm0000022 Clark Gable 1901 1960 actor,soundtrack,producer tt0026752,tt0023382,tt0025316,tt0031381 nm0000023 Judy Garland 1922 1969 soundtrack,actress tt0032138,tt0055031,tt0047522,tt0037059 nm0000024 John Gielgud 1904 2000 actor,writer,director tt0082031,tt0117631,tt0045943,tt0071877 nm0000025 Jerry Goldsmith 1929 2004 music_department,soundtrack,composer tt0117731,tt0112715,tt0077269,tt0119488 nm0000026 Cary Grant 1904 1986 actor,soundtrack,producer tt0034248,tt0048728,tt0053125,tt0056923 nm0000027 Alec Guinness 1914 2000 actor,soundtrack,writer tt0041546,tt0076759,tt0051739,tt0050212 nm0000028 Rita Hayworth 1918 1987 actress,soundtrack,producer tt0036723,tt0040525,tt0046076,tt0038559 nm0000029 Margaux Hemingway 1954 1996 actress,miscellaneous tt0102122,tt0077800,tt0074802,tt0110138 nm0000030 Audrey Hepburn 1929 1993 actress,soundtrack tt0046250,tt0058385,tt0056923,tt0054698 nm0000031 Katharine Hepburn 1907 2003 actress,soundtrack,writer tt0063227,tt0043265,tt0082846,tt0032904 nm0000032 Charlton Heston 1923 2008 actor,director,writer tt0063442,tt0049833,tt0070723,tt0052618 nm0000033 Alfred Hitchcock 1899 1980 director,producer,writer tt0053125,tt0052357,tt0040746,tt0054215 nm0000034 William Holden 1918 1981 actor,soundtrack tt0074958,tt0046359,tt0043014,tt0047437 nm0000035 James Horner 1953 2015 music_department,soundtrack,composer tt0120338,tt0499549,tt0090633,tt0120746 nm0000036 Buster Keaton 1895 1966 actor,writer,director tt0017925,tt0019421,tt0015324,tt0016332 nm0000037 Gene Kelly 1912 1996 soundtrack,actor,miscellaneous tt0053946,tt0041716,tt0045152,tt0043278 nm0000038 Grace Kelly 1929 1982 actress,soundtrack tt0048728,tt0046912,tt0046874,tt0047396 nm0000039 Deborah Kerr 1921 2007 actress,soundtrack tt0039192,tt0049408,tt0050105,tt0045793 nm0000040 Sta

In [0]:
title_akas = spark.read.format("csv") \
                        .option("header", "true") \
                        .option("inferSchema", "true") \
                        .option("delimiter", "\t") \
                        .load("file:/databricks/driver/title.akas.tsv")
title_akas.cache()

Out[14]: DataFrame[titleId: string, ordering: int, title: string, region: string, language: string, types: string, attributes: string, isOriginalTitle: string]

In [0]:
display(title_akas)

titleId ordering title region language types attributes isOriginalTitle tt0000001 1 Карменсіта UA \N imdbDisplay \N 0 tt0000001 2 Carmencita DE \N \N literal title 0 tt0000001 3 Carmencita - spanyol tánc HU \N imdbDisplay \N 0 tt0000001 4 Καρμενσίτα GR \N imdbDisplay \N 0 tt0000001 5 Карменсита RU \N imdbDisplay \N 0 tt0000001 6 Carmencita US \N imdbDisplay \N 0 tt0000001 7 Carmencita \N \N original \N 1 tt0000001 8 カルメンチータ JP ja imdbDisplay \N 0 tt0000002 1 Le clown et ses chiens \N \N original \N 1 tt0000002 2 Le clown et ses chiens FR \N imdbDisplay \N 0 tt0000002 3 A bohóc és kutyái HU \N imdbDisplay \N 0 tt0000002 4 Der Clown und seine Hunde DE \N \N literal title 0 tt0000002 5 Clovnul si cainii sai RO \N imdbDisplay \N 0 tt0000002 6 Клоун и его собаки RU \N imdbDisplay \N 0 tt0000002 7 The Clown and His Dogs US \N \N literal English title 0 tt0000002 8 道化師と犬 JP ja imdbDisplay \N 0 tt0000003 1 Sarmanul Pierrot RO \N imdbDisplay \N 0 tt0000003 2 Szegény Pierrot HU \N imdbDisplay \N 0 tt0000003 3 哀れなピエロ JP ja imdbDisplay \N 0 tt0000003 4 Бідний П'єро UA \N imdbDisplay \N 0 tt0000003 5 Бедный Пьеро RU \N imdbDisplay \N 0 tt0000003 6 Pauvre Pierrot \N \N original \N 1 tt0000003 7 Poor Pierrot GB \N imdbDisplay \N 0 tt0000003 8 Pauvre Pierrot FR \N imdbDisplay \N 0 tt0000003 9 Armer Pierrot DE \N \N literal title 0 tt0000004 1 Un bon bock \N \N original \N 1 tt0000004 2 Un bon bock FR \N imdbDisplay \N 0 tt0000004 3 Ein gutes Glas Bier DE \N \N literal title 0 tt0000004 4 Un ţap de bere RO \N imdbDisplay \N 0 tt0000004 5 Полная кружка пива RU \N imdbDisplay \N 0 tt0000004 6 一杯のビール JP ja imdbDisplay \N 0 tt0000004 7 A Good Beer \N \N \N \N 0 tt0000004 8 Egy jó pohár sör HU \N imdbDisplay \N 0 tt0000005 10 Blacksmith Scene US \N imdbDisplay \N 0 tt0000005 11 Blacksmith Scene \N \N original \N 1 tt0000005 12 The Blacksmith's Forge GB \N \N informal alternative title 0 tt0000005 1 Blacksmithing Scene US \N alternative \N 0 tt0000005 2 Ковальська сцена UA \N imdbDisplay \N 0 tt0000005 3 Сцена в кузне RU \N imdbDisplay \N 0 tt0000005 4 Blacksmith Scene CA en imdbDisplay \N 0 tt0000005 5 Blacksmith Shop GB \N \N informal alternative title 0 tt0000005 6 Blacksmith Scene #1 US \N alternative \N 0 tt0000005 7 Blacksmithing US \N \N informal alternative title 0 tt0000005 8 Schmiedszene DE \N \N literal title 0 tt0000005 9 A patkolókovács HU \N imdbDisplay \N 0 tt0000006 1 Китайский опиумный притон RU \N imdbDisplay \N 0 tt0000006 2 Chinezes Fumadores de Ópio PT \N \N promotional title 0 tt0000006 3 Chinese Opium Den US \N imdbDisplay \N 0 tt0000006 4 Kínai ópiumbarlang HU \N imdbDisplay \N 0 tt0000006 5 Chinese Opium Den \N \N original \N 1 tt0000006 6 Chinesische Opiumhöhle DE \N \N literal title 0 tt0000007 1 Corbett and Courtney Before the Kinetograph US \N imdbDisplay \N 0 tt0000007 2 Corbett and Courtney Before the Kinetograph \N \N original \N 1 tt0000007 3 A Corbett-Courtney bokszmeccs HU \N imdbDisplay \N 0 tt0000007 4 キネトグラフの前のコーベットとコートニー JP ja imdbDisplay \N 0 tt0000007 5 The Corbett-Courtney Fight US \N alternative \N 0 tt0000007 6 Jim Corbett vs. Peter Courtney US \N alternative \N 0 tt0000007 7 Edison Kinetoscopic Record of Boxers \N \N \N \N 0 tt0000007 8 Корбетт и Кортни перед кинематографом RU \N imdbDisplay \N 0 tt0000007 9 Corbett und Courtney vor dem Kinetographen DE \N \N literal title 0 tt0000008 10 Fred Ott's Sneeze \N \N \N \N 0 tt0000008 11 Edison kinetoskopische Aufnahme eines Niesens DE \N \N literal title 0 tt0000008 1 Edison asszisztense tüsszent HU \N imdbDisplay \N 0 tt0000008 2 フレッド・オット－のくしゃみ JP ja imdbDisplay \N 0 tt0000008 3 Edison Kinetoscopic Record of a Sneeze, January 7, 1894 US \N \N complete title 0 tt0000008 4 Edison Kinetoscopic Record of a Sneeze US \N imdbDisplay \N 0 tt0000008 5 Edison Kinetoscopic Record of a Sneeze \N \N original \N 1 tt0000008 6 Fred Otts Niesen DE \N \N literal title 0 tt0000008 7 Чих, записанный на кинетоскоп Эдисона RU \N imdbDisplay \N 0 tt0000008 8 The Films of Thomas A

In [0]:
title_basics = spark.read.format("csv") \
                        .option("header", "true") \
                        .option("inferSchema", "true") \
                        .option("delimiter", "\t") \
                        .load("file:/databricks/driver/title.basics.tsv")
title_basics.cache()

Out[16]: DataFrame[tconst: string, titleType: string, primaryTitle: string, originalTitle: string, isAdult: string, startYear: string, endYear: string, runtimeMinutes: string, genres: string]

In [0]:
display(title_basics)

tconst titleType primaryTitle originalTitle isAdult startYear endYear runtimeMinutes genres tt0000001 short Carmencita Carmencita 0 1894 \N 1 Documentary,Short tt0000002 short Le clown et ses chiens Le clown et ses chiens 0 1892 \N 5 Animation,Short tt0000003 short Pauvre Pierrot Pauvre Pierrot 0 1892 \N 4 Animation,Comedy,Romance tt0000004 short Un bon bock Un bon bock 0 1892 \N 12 Animation,Short tt0000005 short Blacksmith Scene Blacksmith Scene 0 1893 \N 1 Comedy,Short tt0000006 short Chinese Opium Den Chinese Opium Den 0 1894 \N 1 Short tt0000007 short Corbett and Courtney Before the Kinetograph Corbett and Courtney Before the Kinetograph 0 1894 \N 1 Short,Sport tt0000008 short Edison Kinetoscopic Record of a Sneeze Edison Kinetoscopic Record of a Sneeze 0 1894 \N 1 Documentary,Short tt0000009 short Miss Jerry Miss Jerry 0 1894 \N 40 Romance,Short tt0000010 short Leaving the Factory La sortie de l'usine Lumière à Lyon 0 1895 \N 1 Documentary,Short tt0000011 short Akrobatisches Potpourri Akrobatisches Potpourri 0 1895 \N 1 Documentary,Short tt0000012 short The Arrival of a Train L'arrivée d'un train à La Ciotat 0 1896 \N 1 Documentary,Short tt0000013 short The Photographical Congress Arrives in Lyon Le débarquement du congrès de photographie à Lyon 0 1895 \N 1 Documentary,Short tt0000014 short The Waterer Watered L'arroseur arrosé 0 1895 \N 1 Comedy,Short tt0000015 short Autour d'une cabine Autour d'une cabine 0 1894 \N 2 Animation,Short tt0000016 short Boat Leaving the Port Barque sortant du port 0 1895 \N 1 Documentary,Short tt0000017 short Italienischer Bauerntanz Italienischer Bauerntanz 0 1895 \N 1 Documentary,Short tt0000018 short Das boxende Känguruh Das boxende Känguruh 0 1895 \N 1 Short tt0000019 short The Clown Barber The Clown Barber 0 1898 \N \N Comedy,Short tt0000020 short The Derby 1895 The Derby 1895 0 1895 \N 1 Documentary,Short,Sport tt0000022 short Blacksmith Scene Les forgerons 0 1895 \N 1 Documentary,Short tt0000023 short The Sea Baignade en mer 0 1895 \N 1 Documentary,Short tt0000024 short Opening of the Kiel Canal Opening of the Kiel Canal 0 1895 \N \N News,Short tt0000025 short The Oxford and Cambridge University Boat Race The Oxford and Cambridge University Boat Race 0 1895 \N \N News,Short,Sport tt0000026 short The Messers. Lumière at Cards Partie d'écarté 0 1896 \N 1 Documentary,Short tt0000027 short Cordeliers' Square in Lyon Place des Cordeliers à Lyon 0 1895 \N 1 Documentary,Short tt0000028 short Fishing for Goldfish La pêche aux poissons rouges 0 1895 \N 1 Documentary,Short tt0000029 short Baby's Meal Repas de bébé 0 1895 \N 1 Documentary,Short tt0000030 short Rough Sea at Dover Rough Sea at Dover 0 1895 \N 1 Documentary,Short tt0000031 short Jumping the Blanket Le saut à la couverture 0 1895 \N 1 Documentary,Short tt0000032 short Die Serpentintänzerin Die Serpentintänzerin 0 1895 \N 1 Short tt0000033 short Horse Trick Riders La voltige 0 1895 \N 1 Comedy,Documentary,Short tt0000034 short Arrivée d'un train gare de Vincennes Arrivée d'un train gare de Vincennes 0 1896 \N 1 Documentary,Short tt0000035 short Watering the Flowers L'arroseur 0 1896 \N 1 Comedy,Short tt0000036 short Awakening of Rip Awakening of Rip 0 1896 \N 0 Drama,Short tt0000037 short Sea Bathing Baignade en mer 0 1896 \N 1 Short tt0000038 short The Ball Game The Ball Game 0 1898 \N \N Documentary,Short,Sport tt0000039 short Barnet Horse Fair Barnet Horse Fair 0 1896 \N \N Short tt0000040 short Barque sortant du port de Trouville Barque sortant du port de Trouville 0 1896 \N \N Documentary,Short tt0000041 short Bataille de neige Bataille de neige 0 1897 \N 1 Comedy,Documentary,Short tt0000042 short Bateau-mouche sur la Seine Bateau-mouche sur la Seine 0 1896 \N \N Documentary,Short tt0000043 short Batteuse à vapeur Batteuse à vapeur 0 1896 \N \N Documentary,Short tt0000044 short Le bivouac Le bivouac 0 1896 \N 1 Short tt0000045 short Les blanchisseuses Les blanchisseuses 0 1896 \N \N Short tt0000046 short Bois de Boulogne Bois d

In [0]:
'''
title_crew = spark.read.format("csv") \
                        .option("header", "true") \
                        .option("inferSchema", "true") \
                        .option("delimiter", "\t") \
                        .load("file:/databricks/driver/title.crew.tsv")
title_crew.cache()'''

Out[18]: '\ntitle_crew = spark.read.format("csv")                         .option("header", "true")                         .option("inferSchema", "true")                         .option("delimiter", "\t")                         .load("file:/databricks/driver/title.crew.tsv")\ntitle_crew.cache()'

In [0]:
'''
title_episode = spark.read.format("csv") \
                        .option("header", "true") \
                        .option("inferSchema", "true") \
                        .option("delimiter", "\t") \
                        .load("file:/databricks/driver/title.episode.tsv")
title_episode.cache()'''

Out[19]: '\ntitle_episode = spark.read.format("csv")                         .option("header", "true")                         .option("inferSchema", "true")                         .option("delimiter", "\t")                         .load("file:/databricks/driver/title.episode.tsv")\ntitle_episode.cache()'

In [0]:
title_principals = spark.read.format("csv") \
                        .option("header", "true") \
                        .option("inferSchema", "true") \
                        .option("delimiter", "\t") \
                        .load("file:/databricks/driver/title.principals.tsv")
title_principals.cache()

Out[20]: DataFrame[tconst: string, ordering: int, nconst: string, category: string, job: string, characters: string]

In [0]:
display(title_principals)

tconst ordering nconst category job characters tt0000001 1 nm1588970 self \N ["Self"] tt0000001 2 nm0005690 director \N \N tt0000001 3 nm0374658 cinematographer director of photography \N tt0000002 1 nm0721526 director \N \N tt0000002 2 nm1335271 composer \N \N tt0000003 1 nm0721526 director \N \N tt0000003 2 nm1770680 producer producer \N tt0000003 3 nm1335271 composer \N \N tt0000003 4 nm5442200 editor \N \N tt0000004 1 nm0721526 director \N \N tt0000004 2 nm1335271 composer \N \N tt0000005 1 nm0443482 actor \N ["Blacksmith"] tt0000005 2 nm0653042 actor \N ["Assistant"] tt0000005 3 nm0005690 director \N \N tt0000005 4 nm0249379 producer producer \N tt0000006 1 nm0005690 director \N \N tt0000007 1 nm0179163 actor \N \N tt0000007 2 nm0183947 actor \N \N tt0000007 3 nm0005690 director \N \N tt0000007 4 nm0374658 director \N \N tt0000007 5 nm0249379 producer producer \N tt0000008 1 nm0653028 actor \N ["Sneezing Man"] tt0000008 2 nm0005690 director \N \N tt0000008 3 nm0374658 cinematographer \N \N tt0000009 1 nm0063086 actress \N ["Miss Geraldine Holbrook (Miss Jerry)"] tt0000009 2 nm0183823 actor \N ["Mr. Hamilton"] tt0000009 3 nm1309758 actor \N ["Chauncey Depew - the Director of the New York Central Railroad"] tt0000009 4 nm0085156 director \N \N tt0000010 1 nm0525910 director \N \N tt0000011 1 nm3692297 actor \N ["Acrobats"] tt0000011 2 nm0804434 director \N \N tt0000012 1 nm2880396 self \N ["Self"] tt0000012 2 nm9735580 self \N ["Self"] tt0000012 3 nm0525900 self \N ["Self"] tt0000012 4 nm9735581 self \N ["Self"] tt0000012 5 nm0525908 director \N \N tt0000012 6 nm0525910 director \N \N tt0000012 7 nm9735579 self \N ["Self"] tt0000012 8 nm9653419 self \N ["Self"] tt0000013 1 nm0525908 self \N ["Self"] tt0000013 2 nm1715062 self \N ["Self"] tt0000013 3 nm0525910 director \N \N tt0000014 1 nm0166380 actor \N ["The Gardener"] tt0000014 2 nm0244989 actor \N ["The Boy"] tt0000014 3 nm0525910 director \N \N tt0000015 1 nm0721526 director \N \N tt0000016 1 nm0525900 self \N ["Self (on the pier)"] tt0000016 2 nm9735581 self \N ["Self (on the pier)"] tt0000016 3 nm0525910 director \N \N tt0000017 1 nm3691272 actor \N ["The boy"] tt0000017 2 nm3692829 actress \N ["The girl"] tt0000017 3 nm1587194 director \N \N tt0000017 4 nm0804434 director \N \N tt0000018 1 nm3692071 actor \N ["The Boxer"] tt0000018 2 nm0804434 director \N \N tt0000019 1 nm0932055 director \N \N tt0000020 1 nm0010291 director \N \N tt0000020 2 nm0666972 producer producer \N tt0000022 1 nm0525910 director \N \N tt0000023 1 nm0525910 director \N \N tt0000024 1 nm0256651 self \N ["Self - Empress of Gemany"] tt0000024 2 nm0435118 self \N ["Self - Emperor of Germany"] tt0000024 3 nm0010291 director \N \N tt0000024 4 nm0666972 producer producer \N tt0000025 1 nm0010291 director \N \N tt0000025 2 nm0666972 producer producer \N tt0000026 1 nm2350007 actor \N ["Valet"] tt0000026 2 nm0525907 actor \N ["Man playing cards, seated on the left"] tt0000026 3 nm1151424 actor \N ["Man playing cards, seated on the right"] tt0000026 4 nm2354154 actor \N ["Man playing cards, seated in the middle"] tt0000026 5 nm0525910 director \N \N tt0000027 1 nm0525910 director \N \N tt0000028 1 nm2350838 self \N ["Self"] tt0000028 2 nm0525908 self \N ["Self"] tt0000028 3 nm0525910 director \N \N tt0000029 1 nm0525908 self \N ["Self"] tt0000029 2 nm0525900 self \N ["Self"] tt0000029 3 nm2350838 self \N ["Self"] tt0000029 4 nm0525910 director \N \N tt0000030 1 nm0010291 director \N \N tt0000030 2 nm0666972 director \N \N tt0000031 1 nm0525910 director \N \N tt0000032 1 nm3692479 actress \N ["The dancer"] tt0000032 2 nm0804434 director \N \N tt0000033 1 nm0525910 director \N \N tt0000034 1 nm0617588 director \N \N tt0000035 1 nm0617588 director \N \N tt0000036 1 nm0420198 actor \N ["Rip Van Winkle"] tt0000036 2 nm0005690 director \N \N tt0000036 3 nm0410331 writer story "Rip Van Winkle" \N tt0000036 4 nm0005658 cinematographer \N \N tt0000037 1 nm0617588 director \N \N tt0000038 1 nm2960633 self \N ["S

In [0]:

title_ratings = spark.read.format("csv") \
                        .option("header", "true") \
                        .option("inferSchema", "true") \
                        .option("delimiter", "\t") \
                        .load("file:/databricks/driver/title.ratings.tsv")
title_ratings.cache()

Out[22]: DataFrame[tconst: string, averageRating: double, numVotes: int]

In [0]:
display(title_ratings)

tconst averageRating numVotes tt0000001 5.7 1884 tt0000002 5.9 250 tt0000003 6.5 1669 tt0000004 5.8 163 tt0000005 6.2 2490 tt0000006 5.2 166 tt0000007 5.4 776 tt0000008 5.4 2025 tt0000009 5.3 195 tt0000010 6.9 6812 tt0000011 5.3 347 tt0000012 7.4 11708 tt0000013 5.7 1805 tt0000014 7.1 5246 tt0000015 6.2 1005 tt0000016 5.9 1415 tt0000017 4.6 308 tt0000018 5.3 565 tt0000019 5.2 31 tt0000020 4.8 335 tt0000022 5.1 1046 tt0000023 5.7 1365 tt0000024 4.3 91 tt0000025 4.3 42 tt0000026 5.7 1476 tt0000027 5.6 1100 tt0000028 5.1 1017 tt0000029 5.9 3219 tt0000030 5.2 805 tt0000031 5.5 963 tt0000032 5.0 398 tt0000033 5.5 980 tt0000034 5.0 188 tt0000035 5.4 77 tt0000036 4.4 582 tt0000037 4.4 62 tt0000038 4.2 194 tt0000039 3.6 27 tt0000040 4.3 62 tt0000041 6.8 1739 tt0000042 4.2 36 tt0000043 3.6 31 tt0000044 4.1 42 tt0000045 4.1 31 tt0000046 4.0 34 tt0000047 3.7 33 tt0000048 4.9 182 tt0000049 5.1 46 tt0000050 4.1 34 tt0000051 3.8 33 tt0000052 4.4 93 tt0000053 2.9 17 tt0000054 4.9 33 tt0000055 3.8 29 tt0000056 3.2 21 tt0000057 4.6 32 tt0000058 4.3 31 tt0000059 3.6 26 tt0000060 7.5 85 tt0000061 4.1 23 tt0000062 6.4 190 tt0000063 3.6 26 tt0000064 3.4 27 tt0000065 3.6 27 tt0000066 3.4 27 tt0000067 5.7 60 tt0000068 3.9 30 tt0000069 4.9 430 tt0000070 6.4 2551 tt0000071 3.5 28 tt0000072 3.8 26 tt0000073 3.6 29 tt0000074 4.7 30 tt0000075 6.3 1821 tt0000076 4.5 511 tt0000077 4.6 33 tt0000078 3.8 72 tt0000079 4.4 34 tt0000080 4.1 30 tt0000081 4.2 31 tt0000082 4.8 46 tt0000083 4.7 31 tt0000084 4.5 29 tt0000085 4.7 31 tt0000086 4.3 32 tt0000087 4.7 36 tt0000088 4.3 31 tt0000089 6.2 961 tt0000090 4.1 30 tt0000091 6.7 3385 tt0000092 4.7 28 tt0000093 4.8 89 tt0000094 4.8 31 tt0000095 4.8 33 tt0000096 4.5 28 tt0000097 5.0 32 tt0000098 5.1 35 tt0000099 4.7 31 tt0000100 4.7 51 tt0000101 4.7 27 tt0000102 4.8 25 tt0000103 5.4 26 tt0000104 5.6 450 tt0000105 5.0 28 tt0000106 4.5 22 tt0000107 5.2 27 tt0000108 4.4 519 tt0000109 4.5 505 tt0000110 4.4 509 tt0000111 4.5 524 tt0000112 4.5 503 tt0000113 4.3 528 tt0000114 5.1 29 tt0000115 4.8 33 tt0000116 5.1 26 tt0000117 5.2 156 tt0000118 5.1 27 tt0000119 5.4 51 tt0000120 5.3 49 tt0000121 4.3 76 tt0000122 5.1 30 tt0000123 5.7 122 tt0000124 5.1 29 tt0000125 4.8 14 tt0000126 5.5 29 tt0000127 4.6 25 tt0000128 5.6 37 tt0000129 6.5 49 tt0000130 5.5 28 tt0000131 5.8 1315 tt0000132 5.1 1013 tt0000133 6.1 34 tt0000134 4.7 18 tt0000135 6.6 77 tt0000137 4.6 30 tt0000138 6.5 1061 tt0000139 5.3 24 tt0000140 4.9 278 tt0000141 5.2 92 tt0000142 4.7 28 tt0000143 4.9 245 tt0000144 5.1 298 tt0000145 3.9 13 tt0000146 5.1 26 tt0000147 5.2 446 tt0000148 3.7 14 tt0000149 4.6 21 tt0000150 5.4 157 tt0000151 4.9 57 tt0000152 6.1 235 tt0000153 5.7 417 tt0000154 4.1 13 tt0000155 5.1 27 tt0000156 4.1 22 tt0000157 4.1 15 tt0000158 3.5 15 tt0000159 5.3 117 tt0000160 5.3 26 tt0000161 3.9 14 tt0000162 5.4 455 tt0000163 5.1 13 tt0000164 5.0 287 tt0000165 4.6 118 tt0000166 3.8 18 tt0000167 4.0 134 tt0000168 3.5 16 tt0000169 3.8 45 tt0000170 5.2 237 tt0000171 4.5 14 tt0000172 6.1 857 tt0000173 4.9 44 tt0000174 4.6 126 tt0000175 5.0 206 tt0000176 4.9 12 tt0000177 4.6 236 tt0000178 5.2 170 tt0000179 4.5 40 tt0000181 6.8 59 tt0000182 4.9 533 tt0000183 6.0 35 tt0000184 4.9 313 tt0000186 4.2 21 tt0000187 4.9 53 tt0000188 4.0 11 tt0000190 4.9 140 tt0000192 4.6 182 tt0000194 5.4 83 tt0000196 5.2 27 tt0000197 4.4 19 tt0000198 4.4 19 tt0000199 4.3 13 tt0000204 5.7 13 tt0000205 5.0 30 tt0000206 5.0 12 tt0000207 5.5 12 tt0000208 5.7 541 tt0000209 4.7 28 tt0000210 4.9 31 tt0000211 7.5 4098 tt0000213 4.7 227 tt0000214 4.4 93 tt0000215 4.1 121 tt0000218 5.9 616 tt0000219 6.6 30 tt0000222 4.7 225 tt0000223 5.4 18 tt0000225 5.2 35 tt0000228 4.9 265 tt0000229 4.4 106 tt0000230 6.5 1589 tt0000231 3.6 16 tt0000233 6.0 10 tt0000234 6.0 68 tt0000235 6.3 828 tt0000239 5.5 233 tt0000240 4.4 88 tt0000242 6.4 1135 tt0000245 6.2 48 tt0000246 6.4 1008 tt0000247 5.0 472 tt0000248 6.1 1528 tt0000251 5.4 133 tt0000253 4.4 25 tt0000256 4.7 165 tt0000261 5.4 62 tt0000264 5.7 418 tt0000

### Network Inference, Let’s build a network
In the following questions you will look to summarise the data and build a network. We want to examine a network that abstracts how actors and actress are related through their co-participation in movies. To that end perform the following steps:

**Q1** Create a DataFrame that combines **all the information** on each of the titles (i.e., movies, tv-shows, etc …) and **all of the information** the participants in those movies (i.e., actors, directors, etc … ), make sure the actual names of the movies and participants are included. It may be worth reviewing the following questions to see how this dataframe will be used.

How many rows does your dataframe have?

In [0]:
#using only 3 tables, the ones needed for the following questions and performing two left joins to merge these tables. i choose to use left join because we want ALL the information of each title (so ALL titles) and only the information of the participats in THOSE titles. title_principals is the table between title_basics and name_basics that connect both.

In [0]:
dataframe = title_basics.join(title_principals, on='tconst',how='left')\
                        .join(name_basics, on='nconst', how='left').cache()

In [0]:
display(dataframe)

nconst tconst titleType primaryTitle originalTitle isAdult startYear endYear runtimeMinutes genres ordering category job characters primaryName birthYear deathYear primaryProfession knownForTitles nm0000086 tt0083109 movie The Cabbage Soup La soupe aux choux 0 1981 \N 98 Comedy,Drama,Sci-Fi 1 actor \N ["Claude Ratinier (Le Glaude)"] Louis de Funès 1914 1983 actor,writer,soundtrack tt0074103,tt0079200,tt0069747,tt0064425 nm0000086 tt0062120 movie The Little Bather Le petit baigneur 0 1968 \N 96 Comedy 1 actor \N ["Louis-Philippe Fourchaume"] Louis de Funès 1914 1983 actor,writer,soundtrack tt0074103,tt0079200,tt0069747,tt0064425 nm0000086 tt2949354 tvMovie Monsieur de Funès Monsieur de Funès 0 2013 \N 82 Documentary 1 archive_footage \N ["Self"] Louis de Funès 1914 1983 actor,writer,soundtrack tt0074103,tt0079200,tt0069747,tt0064425 nm0000086 tt0057422 movie Squeak-squeak Pouic-Pouic 0 1963 \N 86 Comedy 1 actor \N ["Léonard Monestier"] Louis de Funès 1914 1983 actor,writer,soundtrack tt0074103,tt0079200,tt0069747,tt0064425 nm0000086 tt6983458 tvEpisode Chroniques de France N° 1 Chroniques de France N° 1 0 1964 \N 26 Documentary 3 self \N ["Self - (segment \"Maurice Chevalier\")"] Louis de Funès 1914 1983 actor,writer,soundtrack tt0074103,tt0079200,tt0069747,tt0064425 nm0000086 tt16374010 tvEpisode Episode dated 16 June 1963 Episode dated 16 June 1963 0 1963 \N \N News 3 self \N ["Self"] Louis de Funès 1914 1983 actor,writer,soundtrack tt0074103,tt0079200,tt0069747,tt0064425 nm0000086 tt0043691 tvMovie The Gamblers Les joueurs 0 1950 \N 50 Comedy 1 actor \N ["Piotr Petrovitch"] Louis de Funès 1914 1983 actor,writer,soundtrack tt0074103,tt0079200,tt0069747,tt0064425 nm0000086 tt0335439 short Station mondaine Station mondaine 0 1951 \N \N Documentary,Short 1 actor \N ["Le narrateur"] Louis de Funès 1914 1983 actor,writer,soundtrack tt0074103,tt0079200,tt0069747,tt0064425 nm0000086 tt2223153 tvEpisode Episode dated 14 May 1966 Episode dated 14 May 1966 0 1966 \N \N Documentary 6 self \N ["Self"] Louis de Funès 1914 1983 actor,writer,soundtrack tt0074103,tt0079200,tt0069747,tt0064425 nm0000086 tt2288018 tvEpisode Le Corniaud, Tous les enfants du monde Le Corniaud, Tous les enfants du monde 0 1965 \N 43 Documentary 2 self \N ["Self"] Louis de Funès 1914 1983 actor,writer,soundtrack tt0074103,tt0079200,tt0069747,tt0064425 nm0000086 tt5798854 tvEpisode Episode dated 15 March 1980 Episode dated 15 March 1980 0 1980 \N \N \N 2 self \N ["Self"] Louis de Funès 1914 1983 actor,writer,soundtrack tt0074103,tt0079200,tt0069747,tt0064425 nm0000086 tt12069126 tvEpisode La science des réseaux sociaux La science des réseaux sociaux 0 2020 \N \N Documentary 8 archive_footage \N ["Self"] Louis de Funès 1914 1983 actor,writer,soundtrack tt0074103,tt0079200,tt0069747,tt0064425 nm0000086 tt0058089 movie Fantomas Fantômas 0 1964 \N 104 Adventure,Comedy,Crime 2 actor \N ["Le commissaire Juve"] Louis de Funès 1914 1983 actor,writer,soundtrack tt0074103,tt0079200,tt0069747,tt0064425 nm0000086 tt4084302 tvEpisode Louis de Funès Louis de Funès 0 2010 \N \N Documentary 2 archive_footage \N ["Self"] Louis de Funès 1914 1983 actor,writer,soundtrack tt0074103,tt0079200,tt0069747,tt0064425 nm0000086 tt7415870 tvMovie La servante La servante 0 1953 \N \N \N 1 actor \N ["Le maître de maison"] Louis de Funès 1914 1983 actor,writer,soundtrack tt0074103,tt0079200,tt0069747,tt0064425 nm0000086 tt14979878 tvMovie Louis de Funès, Créature/Créateur Louis de Funès, Créature/Créateur 0 2020 \N 52 Documentary 1 archive_footage \N \N Louis de Funès 1914 1983 actor,writer,soundtrack tt0074103,tt0079200,tt0069747,tt0064425 nm0000086 tt5815870 tvEpisode Episode dated 15 November 1970 Episode dated 15 November 1970 0 1970 \N 44 \N 2 self \N ["Self"] Louis de Funès 1914 1983 actor,writer,soundtrack tt0074103,tt0079200,tt0069747,tt0064425 nm0000086 tt0067274 movie Jo Jo 0 1971 \N 85 Comedy,Crime,Thriller 1 actor \N ["Antoine Brisebard"] Louis de Funès 1914 1983 actor,writer,soundtr

My final anwser is:

In [0]:
dataframe.count()

Out[27]: 51545700

**Q2** Create a new DataFrame based on the previous step, with the following removed:
1. Any participant that is not an actor or actress (as measured by the category column);
1. All adult movies;
1. All dead actors or actresses;
1. All actors or actresses born before 1920 or with no date of birth listed;
1. All titles that are not of the type movie.

How many rows does your dataframe have?

In [0]:
#filtering the dataframe according with the constraints 

In [0]:
df = dataframe.filter(((dataframe.category == 'actor') | (dataframe.category == 'actress')) &\
                      (dataframe.isAdult == 0)& \
                      (dataframe.deathYear == "\\N") & \
                      ((dataframe.birthYear >= 1920) & \
                      (dataframe.birthYear != '\\N')) & \
                      (dataframe.titleType == 'movie' )).cache()

In [0]:
display(df)

nconst tconst titleType primaryTitle originalTitle isAdult startYear endYear runtimeMinutes genres ordering category job characters primaryName birthYear deathYear primaryProfession knownForTitles nm0000198 tt0110116 movie Immortal Beloved Immortal Beloved 0 1994 \N 121 Biography,Drama,Music 1 actor \N ["Ludwig van Beethoven"] Gary Oldman 1958 \N actor,soundtrack,producer tt0468569,tt1340800,tt0103874,tt4555426 nm0000198 tt1345836 movie The Dark Knight Rises The Dark Knight Rises 0 2012 \N 164 Action,Drama 4 actor \N ["Commissioner Gordon"] Gary Oldman 1958 \N actor,soundtrack,producer tt0468569,tt1340800,tt0103874,tt4555426 nm0000198 tt0097125 movie Criminal Law Criminal Law 0 1988 \N 117 Crime,Drama,Thriller 1 actor \N ["Ben Chase"] Gary Oldman 1958 \N actor,soundtrack,producer tt0468569,tt1340800,tt0103874,tt4555426 nm0000198 tt3239932 movie True Crime: The Movie True Crime: The Movie 0 2012 \N 87 Documentary 3 actor \N ["Rocky","Agent Masterson"] Gary Oldman 1958 \N actor,soundtrack,producer tt0468569,tt1340800,tt0103874,tt4555426 nm0000198 tt0208874 movie The Contender The Contender 0 2000 \N 126 Drama,Thriller 2 actor \N ["Shelly Runyon"] Gary Oldman 1958 \N actor,soundtrack,producer tt0468569,tt1340800,tt0103874,tt4555426 nm0000198 tt0100519 movie Rosencrantz & Guildenstern Are Dead Rosencrantz & Guildenstern Are Dead 0 1990 \N 117 Comedy,Drama 1 actor \N ["Rosencrantz"] Gary Oldman 1958 \N actor,soundtrack,producer tt0468569,tt1340800,tt0103874,tt4555426 nm0000198 tt6111574 movie The Woman in the Window The Woman in the Window 0 2021 \N 100 Crime,Drama,Mystery 2 actor \N ["Alistair Russell"] Gary Oldman 1958 \N actor,soundtrack,producer tt0468569,tt1340800,tt0103874,tt4555426 nm0000198 tt0118571 movie Air Force One Air Force One 0 1997 \N 124 Action,Drama,Thriller 2 actor \N ["Ivan Korshunov"] Gary Oldman 1958 \N actor,soundtrack,producer tt0468569,tt1340800,tt0103874,tt4555426 nm0000198 tt2103281 movie Dawn of the Planet of the Apes Dawn of the Planet of the Apes 0 2014 \N 130 Action,Adventure,Drama 1 actor \N ["Dreyfus"] Gary Oldman 1958 \N actor,soundtrack,producer tt0468569,tt1340800,tt0103874,tt4555426 nm0000198 tt10618286 movie Mank Mank 0 2020 \N 131 Biography,Comedy,Drama 1 actor \N ["Herman Mankiewicz"] Gary Oldman 1958 \N actor,soundtrack,producer tt0468569,tt1340800,tt0103874,tt4555426 nm0000198 tt2461520 movie Man Down Man Down 0 2015 \N 90 Drama,Thriller,War 3 actor \N ["Counselor Peyton"] Gary Oldman 1958 \N actor,soundtrack,producer tt0468569,tt1340800,tt0103874,tt4555426 nm0000198 tt0119116 movie The Fifth Element The Fifth Element 0 1997 \N 126 Action,Adventure,Sci-Fi 3 actor \N ["Zorg"] Gary Oldman 1958 \N actor,soundtrack,producer tt0468569,tt1340800,tt0103874,tt4555426 nm0000198 tt0120738 movie Lost in Space Lost in Space 0 1998 \N 130 Action,Adventure,Family 1 actor \N ["Dr. Smith","Spider Smith"] Gary Oldman 1958 \N actor,soundtrack,producer tt0468569,tt1340800,tt0103874,tt4555426 nm0000198 tt0307466 movie Sin Sin 0 2003 \N 107 Crime,Drama,Thriller 1 actor \N ["Charlie Strom"] Gary Oldman 1958 \N actor,soundtrack,producer tt0468569,tt1340800,tt0103874,tt4555426 nm0000198 tt1037705 movie The Book of Eli The Book of Eli 0 2010 \N 118 Action,Adventure,Drama 4 actor \N ["Carnegie"] Gary Oldman 1958 \N actor,soundtrack,producer tt0468569,tt1340800,tt0103874,tt4555426 nm0000198 tt1626175 movie One Night in Turin One Night in Turin 0 2010 \N 90 Documentary,Sport 1 actor \N ["Narrator"] Gary Oldman 1958 \N actor,soundtrack,producer tt0468569,tt1340800,tt0103874,tt4555426 nm0000198 tt5865326 movie The Laundromat The Laundromat 0 2019 \N 96 Comedy,Crime,Drama 1 actor \N ["Jürgen Mossack"] Gary Oldman 1958 \N actor,soundtrack,producer tt0468569,tt1340800,tt0103874,tt4555426 nm0000198 tt0110413 movie Léon: The Professional Léon 0 1994 \N 110 Action,Crime,Drama 2 actor \N ["Stansfield"] Gary Oldman 1958 \N actor,soundtrack,producer tt0468569,tt1340800,tt0103874,tt4555426 nm0000198 tt1067106 movie A Christmas C

My final anwser is:

In [0]:
df.count()

Out[31]: 487603

**Q3** Convert the above Dataframe to an RDD. Use map and reduce to create a paired RDD which counts how many movies each actor / actress appears in.

Display names of the top 10 actors/actresses according to the number of movies in which they appeared. Be careful to deal with different actors / actresses with the same name, these could be different people.

In [0]:
#First, converting the dataframe to an rdd and using not only the name (which we want to display) but also the 'nconst' (which is unique).

In [0]:
paired_RDD = df.rdd.map(lambda x: ((x[0], x[14]), 1))
count_RDD = paired_RDD.reduceByKey(lambda a,b : a + b)
sorted_RDD = count_RDD.sortBy(lambda x: x[1], ascending=False)

My final anwser is:

In [0]:
top_10 = sorted_RDD.map(lambda x: (x[0][1], x[1]))
top_10.take(10)

Out[34]: [('Brahmanandam', 796),
 ('Mammootty', 364),
 ('Mohanlal', 351),
 ('Mithun Chakraborty', 329),
 ('Shakti Kapoor', 322),
 ('Eric Roberts', 313),
 ('Cüneyt Arkin', 300),
 ('Jagathi Sreekumar', 293),
 ('Dharmendra', 270),
 ('Helen', 269)]

**Q4** Start with the dataframe from Q2. Generate a DataFrame that lists all links of your network. Here we shall consider that a link connects a pair of actors/actresses if they participated in at least one movie together (actors / actresses should be represented by their unique ID's). For every link we then need anytime a pair of actors were together in a movie as a link in each direction (A -> B and B -> A). However links should be distinct we do not need duplicates when two actors worked together in several movies.

In [0]:
#creating the edges

#as the question only ask for the links (edges) i will only create the vertices below on page rank question

In [0]:
df_edges = (df.select(df['nconst'].alias('n1'),'tconst')).join(df.select(df['nconst'].alias('n2'),'tconst'), on='tconst')
df_filtered = df_edges.filter(df_edges.n1 != df_edges.n2).cache()    # to guarantee that n1 and n2 have to be different


In [0]:
# we can even check that when A -> B occur also occur B -> A, i mean, when n1 worked with n2 in a movie, n2 also worked with n1 in the same movie. 
#Example:

In [0]:
df_filtered.filter((df_filtered.n1 == 'nm0000113') & (df_filtered.n2 == 'nm0358316')|\
                    (df_filtered.n2 == 'nm0000113') & (df_filtered.n1 == 'nm0358316')).display()

tconst n1 n2 tt2293640 nm0000113 nm0358316 tt2293640 nm0358316 nm0000113

In [0]:
#now i need to guarantee that links should be distinct i.e when two actors worked together in several movies.

In [0]:
EdgesListRow = df_filtered.select('n1','n2').distinct().collect()
sourceColumn = StructField('src', StringType(),True)
destinationColumn = StructField('dst', StringType(), True)
edgeSchema = StructType([sourceColumn, destinationColumn])
edgeDataFrame = sqlContext.createDataFrame(EdgesListRow, edgeSchema).persist()

My final anwser is:

In [0]:
edgeDataFrame.display()

src dst nm0051880 nm0502262 nm0343857 nm0092573 nm0001696 nm0507659 nm0009629 nm0796396 nm0516272 nm0009629 nm0148661 nm0811660 nm0001166 nm0329640 nm0012361 nm2588787 nm0560364 nm0004760 nm0000800 nm0004760 nm0961737 nm0620636 nm1099171 nm0793402 nm2432221 nm0004133 nm0749263 nm0004778 nm0508293 nm0066080 nm0817563 nm1172256 nm2392521 nm3856284 nm0570593 nm0578853 nm7709590 nm0865411 nm1328476 nm7745464 nm2142336 nm0378245 nm0033843 nm1299540 nm3460467 nm0383263 nm5737604 nm4023829 nm0731075 nm2002649 nm4110608 nm0958790 nm0223213 nm5670230 nm1533235 nm1106590 nm2204070 nm2964838 nm0002071 nm0000154 nm0823828 nm3906069 nm0352020 nm0917076 nm0756378 nm0747155 nm0498103 nm0289710 nm0395830 nm0895249 nm0001737 nm0330687 nm0297956 nm0619185 nm0644570 nm0039591 nm0149615 nm0789497 nm1250829 nm1253508 nm0070847 nm0356335 nm0083534 nm0149822 nm0060607 nm0788881 nm0578853 nm1168915 nm3945531 nm4392880 nm1690331 nm5411494 nm1553954 nm1389342 nm0924210 nm0000158 nm2007030 nm2002649 nm0438092 nm0007113 nm0000852 nm0094789 nm0959113 nm0508085 nm0470268 nm0877784 nm0457586 nm0906723 nm0304801 nm0000702 nm0000205 nm0000103 nm0081221 nm0278528 nm0515123 nm0005029 nm0028265 nm2417479 nm0476483 nm3350260 nm1486061 nm0078890 nm1347988 nm1922655 nm0693871 nm1431643 nm0515296 nm0115161 nm0765597 nm3948952 nm0902188 nm0887696 nm0000238 nm0002743 nm0001923 nm0000896 nm0197240 nm0497642 nm0563147 nm0537648 nm0261164 nm0009629 nm0004787 nm0448204 nm0239764 nm0171811 nm1105443 nm0810653 nm0608090 nm0931329 nm4161568 nm2161450 nm0671032 nm0000295 nm1346956 nm0029875 nm1604999 nm0501845 nm2202521 nm2113653 nm0222643 nm2404933 nm0225944 nm2568192 nm1403892 nm1507882 nm0000663 nm0606637 nm1863960 nm2017262 nm2204070 nm2415604 nm0850162 nm1161078 nm0621497 nm0627549 nm0546391 nm0379464 nm0156467 nm0948154 nm0284533 nm0497920 nm0200702 nm0222922 nm0644461 nm0255828 nm0001399 nm0240565 nm0019815 nm0059631 nm0014053 nm0510054 nm0611663 nm0944834 nm0475982 nm1083116 nm0957513 nm0233616 nm1505499 nm0947236 nm0942482 nm0000279 nm0668271 nm1172256 nm2855554 nm1227814 nm0000342 nm0671032 nm0671767 nm0613147 nm0835604 nm4072308 nm3945531 nm0837421 nm4429739 nm7308187 nm0902028 nm0225944 nm2016364 nm0042305 nm0901549 nm7548342 nm5473782 nm0000323 nm0858573 nm0385032 nm0490521 nm0415777 nm0763278 nm0051706 nm0667271 nm0474400 nm0594909 nm0490513 nm0001599 nm0000995 nm0289098 nm0226820 nm0043362 nm0059631 nm0612992 nm0768614 nm0000459 nm0004742 nm1027949 nm0720671 nm1833459 nm2063642 nm0508293 nm1687404 nm2070336 nm1851377 nm9864957 nm0001595 nm1842974 nm0000205 nm0417310 nm1917418 nm0000704 nm3994408 nm0938754 nm0231702 nm0158306 nm4045867 nm0049218 nm0229301 nm5563056 nm0125637 nm13406367 nm0893142 nm2207617 nm1533235 nm2964838 nm2204070 nm0850162 nm2204070 nm0518715 nm2803937 nm0005443 nm2197298 nm0354795 nm2597881 nm0094789 nm0000852 nm0000668 nm0001737 nm0001609 nm0001399 nm0001298 nm0000532 nm0222126 nm0088798 nm0577935 nm0437708 nm0000821 nm0802080 nm0084443 nm0611663 nm0126284 nm0001015 nm1631616 nm1763745 nm0606124 nm1748835 nm1522340 nm1291022 nm0933988 nm2061362 nm0742655 nm1192254 nm0944624 nm1328476 nm0628877 nm0910709 nm1008967 nm1854680 nm1066117 nm1008063 nm0001477 nm0424559 nm0169480 nm0000276 nm1244980 nm0344736 nm0000238 nm0005511 nm3856284 nm3085379 nm3007024 nm3158652 nm0580055 nm1378553 nm2118328 nm2202742 nm0620617 nm0457556 nm2431533 nm5397442 nm0005346 nm0000803 nm1640326 nm2534923 nm4780813 nm1279870 nm6819854 nm3886028 nm1002218 nm1503722 nm0079354 nm0191520 nm0274088 nm0073043 nm0000422 nm0168554 nm0507659 nm0000115 nm0516225 nm0400606 nm0289098 nm0036523 nm0644570 nm0847562 nm0374974 nm0347901 nm0261886 nm0309676 nm2403277 nm0893941 nm1142921 nm0710938 nm0850850 nm1499518 nm0411017 nm0407514 nm3279025 nm0944624 nm1357493 nm1328476 nm0327076 nm0207338 nm0048081 nm0229301 nm0008346 nm0620617 nm1281073 nm1863960 nm2600157 nm2197080 nm0437630 nm0415777 nm0834479 nm0487884 nm0000902 nm0942646 nm0486661 nm0928708 nm0395830 nm0667271 nm0027810 n

**Q5** Compute the page rank of each actor. This can be done using GraphFrames or
by using RDDs and the iterative implementation of the PageRank algorithm. Do not take
more than 5 iterations and use reset probility = 0.1.

List the top 10 actors / actresses by pagerank.

In [0]:
#creating the vertices, knowing that they must be unique so we need to drop duplicated ids.

In [0]:
verticesListRow = df.select('nconst').dropDuplicates().collect()
verticesSchema = StructType([StructField('id', StringType(), True)])
verticesDataFrame = sqlContext.createDataFrame(verticesListRow, verticesSchema).persist()

In [0]:
verticesDataFrame.display()

id nm0000198 nm0000354 nm0000362 nm0000767 nm0002222 nm0003678 nm0004709 nm0004763 nm0005037 nm0005258 nm0005349 nm0005535 nm0008161 nm0014306 nm0015901 nm0016966 nm0017118 nm0018808 nm0027424 nm0032810 nm0035700 nm0041436 nm0046770 nm0048122 nm0053910 nm0055474 nm0057029 nm0057741 nm0058547 nm0064609 nm0066984 nm0068464 nm0068551 nm0076427 nm0082732 nm0084901 nm0090120 nm0091035 nm0091379 nm0092283 nm0100056 nm0102195 nm0102522 nm0104427 nm0113463 nm0114931 nm0118840 nm0120390 nm0121141 nm0126962 nm0132436 nm0134743 nm0136399 nm0145579 nm0147022 nm0148430 nm0155923 nm0156741 nm0162560 nm0165145 nm0165411 nm0166388 nm0166394 nm0171627 nm0186569 nm0191412 nm0201956 nm0202497 nm0212204 nm0213142 nm0214302 nm0215229 nm0217529 nm0225598 nm0225777 nm0227862 nm0235577 nm0237296 nm0238337 nm0246570 nm0246820 nm0252434 nm0259074 nm0266235 nm0276150 nm0277932 nm0279996 nm0280071 nm0280355 nm0286925 nm0287055 nm0293255 nm0303177 nm0303599 nm0305342 nm0311343 nm0315321 nm0320759 nm0321587 nm0328601 nm0330879 nm0335916 nm0339034 nm0340952 nm0350453 nm0352203 nm0352921 nm0358866 nm0360305 nm0361013 nm0363146 nm0367636 nm0369213 nm0380489 nm0394123 nm0398700 nm0398907 nm0399466 nm0405966 nm0407661 nm0410066 nm0413824 nm0421105 nm0424582 nm0427274 nm0435323 nm0439797 nm0443365 nm0446705 nm0459031 nm0462319 nm0463100 nm0466119 nm0467379 nm0468600 nm0474429 nm0481714 nm0481731 nm0482652 nm0490352 nm0494066 nm0505949 nm0518216 nm0522844 nm0535702 nm0539605 nm0549123 nm0549526 nm0549603 nm0585651 nm0593924 nm0598201 nm0600488 nm0601105 nm0602000 nm0602383 nm0605510 nm0606396 nm0608026 nm0608067 nm0616054 nm0622637 nm0628054 nm0635108 nm0640669 nm0641067 nm0644817 nm0647303 nm0664440 nm0668241 nm0671211 nm0678959 nm0681193 nm0682893 nm0690074 nm0696810 nm0699132 nm0703352 nm0706537 nm0707174 nm0710490 nm0710769 nm0719773 nm0727382 nm0733586 nm0738722 nm0739705 nm0744776 nm0753922 nm0758318 nm0759437 nm0764227 nm0766307 nm0769038 nm0770773 nm0788120 nm0792354 nm0793248 nm0796250 nm0798043 nm0803370 nm0812556 nm0812926 nm0821656 nm0836752 nm0837476 nm0841301 nm0842295 nm0847128 nm0847265 nm0854911 nm0857855 nm0863599 nm0867959 nm0876255 nm0877270 nm0892471 nm0895005 nm0897197 nm0900511 nm0900644 nm0904886 nm0911320 nm0919258 nm0919442 nm0922751 nm0927952 nm0928422 nm0929720 nm0931158 nm0933264 nm0935768 nm0943671 nm0947166 nm0948058 nm0990253 nm0998254 nm1005239 nm1006993 nm1009731 nm10102770 nm1016351 nm10180701 nm1027678 nm1032842 nm10365133 nm1042518 nm1043479 nm10529651 nm10644582 nm1065909 nm1067213 nm1072877 nm1077542 nm1081300 nm1085024 nm1099623 nm1101768 nm1107941 nm1115068 nm1117469 nm1127988 nm11289314 nm1130777 nm11344879 nm11419217 nm1148917 nm11515428 nm1151747 nm1162399 nm1167195 nm1167592 nm1167655 nm1172478 nm1173300 nm1174035 nm1175940 nm11770482 nm1204465 nm1208116 nm1213235 nm1216597 nm1227181 nm1228340 nm1246512 nm1246742 nm1249372 nm1296585 nm1315225 nm1327092 nm1327166 nm1327219 nm13297966 nm1346793 nm1366703 nm1367819 nm1372250 nm1376193 nm1384462 nm1391390 nm1391994 nm1393792 nm1415658 nm1441714 nm1462363 nm1480555 nm1485036 nm1486647 nm1492443 nm1492784 nm1514484 nm1515028 nm1517148 nm1522629 nm1530895 nm1542592 nm1559179 nm1563338 nm1564283 nm1588112 nm1594301 nm1616937 nm1620510 nm1621849 nm1628032 nm1644138 nm1645878 nm1652431 nm1654844 nm1669979 nm1671380 nm1671717 nm1678574 nm1719883 nm1721707 nm1726021 nm1726927 nm1727323 nm1759049 nm1814092 nm1816263 nm1822793 nm1824148 nm1843960 nm1851772 nm1938397 nm1966962 nm1968316 nm1987473 nm1996465 nm1997065 nm1997609 nm2007628 nm2056385 nm2079368 nm2103508 nm2107073 nm2115926 nm2124052 nm2162912 nm2200825 nm2244829 nm2267821 nm2283410 nm2309449 nm2325108 nm2332146 nm2398362 nm2418386 nm2487121 nm2518213 nm2522536 nm2554701 nm2570728 nm2612218 nm2616518 nm2620667 nm2645828 nm2653910 nm2658331 nm2658780 nm2726391 nm2731972 nm2748752 nm2751858 nm2755731 nm2760504 nm2807269 nm2822547 nm2847302 nm2857804 nm2907715 nm2932082 nm2934031 nm2934652 nm2980797 nm3016599 nm3018325 nm3020

In [0]:
# creating a graph and then applying the page rank algorithm

In [0]:
ourGraph = GraphFrame(verticesDataFrame, edgeDataFrame)
pageRanks = ourGraph.pageRank(resetProbability=0.1, maxIter = 5)
pageRank_vertices = pageRanks.vertices.sort('pagerank',ascending = False)

My final answear is:

In [0]:
pageRank_vertices.show(10)

+---------+------------------+
|       id|          pagerank|
+---------+------------------+
|nm0000616| 53.69072773571389|
|nm0000514|28.450223793553228|
|nm0001744|27.191574641528177|
|nm0001803| 24.95031828455196|
|nm0920460|21.712645095656526|
|nm0000448| 21.46497421351268|
|nm0001595|21.208161365875714|
|nm0001698| 19.77463867014523|
|nm0000367|19.645853356534897|
|nm0004193|19.171056041079787|
+---------+------------------+
only showing top 10 rows



**Q6**: Create an RDD with the number of outDegrees for each actor. Display the top 10 by outdegrees.

My final answear is:

In [0]:
RDD_outDegrees = ourGraph.outDegrees.rdd.map(lambda x: (x[0],x[1]))\
                                        .sortBy(lambda x: x[1], ascending=False)
RDD_outDegrees.take(10)

Out[48]: [('nm0000616', 486),
 ('nm0000514', 284),
 ('nm0001744', 265),
 ('nm0000367', 264),
 ('nm0945189', 257),
 ('nm0256628', 253),
 ('nm0451600', 252),
 ('nm0001803', 246),
 ('nm0149822', 240),
 ('nm0004109', 234)]

### Let’s play Kevin’s own game

**Q7** Start with the graphframe / dataframe you developed in the previous questions. Using Spark GraphFrame and/or Spark Core library perform the following steps:

1. Identify the id of Kevin Bacon, there are two actors named ‘Kevin Bacon’, we will use the one with the highest degree, that is, the one that participated in most titles;
1. Estimate the shortest path between every actor in the database actors and Kevin Bacon, keep a dataframe with this information as you will need it later;
1. Summarise the data, that is, count the number of actors at each number of degress from kevin bacon (you will need to deal with actors unconnected to kevin bacon, if not connected to Kevin Bacon given these actors / actresses a score/degree of 20).

In [0]:
# 1.

# first lets get the IDs of the actors 'Kevin Bacon'; in this exercise i only got one Kevin Bacon 

In [0]:
kevin_bacon = df.filter(df['primaryName'] == 'Kevin Bacon')\
             .groupBy('nconst')\   # just to summarize the information
             .count()\
             .show()

+---------+-----+
|   nconst|count|
+---------+-----+
|nm0000102|   52|
+---------+-----+



In [0]:
id_kevin = ourGraph.outDegrees.filter(ourGraph.outDegrees.id == 'nm0000102') \
                   .sort('outDegree', ascendig = False)\    # see note below
                   .select('id').collect()[0]['id']

id_kevin
# this code was done taking into consideraion that i was supposed to have two 'Kevin Bacon' and choose the one with the highest degree. the only thing missing is that if i really had two 'Kevin Bacon' i would need: filter((ourGraph.outDegrees.id == 'nm0000102')|(ourGraph.outDegrees.id == 'the second id')) and then the sort makes sense.

Out[51]: 'nm0000102'

In [0]:
#2.

shortest_path = ourGraph.shortestPaths(landmarks=[id_kevin]).cache()

In [0]:
shortest_path.display()

id distances nm0057029 Map(nm0000102 -> 4) nm0064609 Map() nm8760959 Map() nm0253708 Map(nm0000102 -> 2) nm2916662 Map(nm0000102 -> 3) nm1597349 Map() nm0057080 Map(nm0000102 -> 4) nm1335115 Map(nm0000102 -> 4) nm3135033 Map(nm0000102 -> 6) nm0177208 Map(nm0000102 -> 4) nm2601709 Map(nm0000102 -> 4) nm3052894 Map(nm0000102 -> 4) nm0156697 Map(nm0000102 -> 3) nm0212563 Map(nm0000102 -> 2) nm1864689 Map() nm7193146 Map(nm0000102 -> 3) nm0107175 Map(nm0000102 -> 3) nm2825922 Map(nm0000102 -> 5) nm1321029 Map(nm0000102 -> 3) nm3859362 Map(nm0000102 -> 4) nm7075530 Map() nm0302462 Map() nm0482057 Map(nm0000102 -> 4) nm1863784 Map(nm0000102 -> 4) nm0510810 Map(nm0000102 -> 4) nm2366503 Map() nm4059532 Map(nm0000102 -> 3) nm4764807 Map(nm0000102 -> 3) nm0001796 Map(nm0000102 -> 3) nm0910180 Map(nm0000102 -> 3) nm1745216 Map(nm0000102 -> 2) nm0000300 Map(nm0000102 -> 2) nm0037410 Map(nm0000102 -> 3) nm0344094 Map(nm0000102 -> 5) nm0719800 Map(nm0000102 -> 4) nm2393332 Map() nm0389333 Map(nm0000102 -> 2) nm2020015 Map(nm0000102 -> 5) nm0463492 Map(nm0000102 -> 4) nm2598542 Map() nm0543212 Map(nm0000102 -> 6) nm3675845 Map() nm0641505 Map(nm0000102 -> 3) nm1057742 Map(nm0000102 -> 5) nm1505005 Map(nm0000102 -> 5) nm1724742 Map(nm0000102 -> 4) nm2509006 Map(nm0000102 -> 3) nm6130752 Map() nm0618123 Map(nm0000102 -> 5) nm8910087 Map(nm0000102 -> 5) nm0052103 Map(nm0000102 -> 4) nm1185383 Map(nm0000102 -> 4) nm0196230 Map(nm0000102 -> 3) nm0732033 Map() nm1330692 Map(nm0000102 -> 5) nm13466150 Map() nm1855801 Map(nm0000102 -> 4) nm0029775 Map(nm0000102 -> 4) nm1686536 Map() nm0142000 Map(nm0000102 -> 3) nm8429968 Map(nm0000102 -> 5) nm1840504 Map(nm0000102 -> 3) nm3241429 Map() nm0095966 Map(nm0000102 -> 5) nm3306913 Map(nm0000102 -> 4) nm1342026 Map(nm0000102 -> 3) nm0687913 Map(nm0000102 -> 4) nm1312105 Map(nm0000102 -> 3) nm2567228 Map(nm0000102 -> 4) nm3795370 Map(nm0000102 -> 4) nm0820883 Map(nm0000102 -> 5) nm2379659 Map() nm0649638 Map(nm0000102 -> 4) nm1840763 Map() nm2505094 Map(nm0000102 -> 5) nm7443364 Map(nm0000102 -> 4) nm0797580 Map() nm1377079 Map(nm0000102 -> 2) nm2336072 Map(nm0000102 -> 3) nm7766905 Map() nm0088008 Map() nm0745412 Map(nm0000102 -> 4) nm0255820 Map(nm0000102 -> 4) nm0719869 Map(nm0000102 -> 4) nm3109658 Map(nm0000102 -> 3) nm1446959 Map() nm0000188 Map(nm0000102 -> 2) nm1366311 Map(nm0000102 -> 3) nm2005912 Map() nm0153688 Map() nm2237905 Map(nm0000102 -> 4) nm0137585 Map(nm0000102 -> 3) nm0148701 Map() nm0441128 Map(nm0000102 -> 3) nm0631761 Map(nm0000102 -> 3) nm0823507 Map(nm0000102 -> 3) nm4149045 Map() nm6807858 Map(nm0000102 -> 2) nm0491006 Map(nm0000102 -> 4) nm5683678 Map(nm0000102 -> 4) nm0029027 Map() nm0772958 Map() nm0313621 Map(nm0000102 -> 4) nm0127716 Map(nm0000102 -> 4) nm0580669 Map(nm0000102 -> 3) nm5201678 Map() nm8306570 Map(nm0000102 -> 4) nm8905347 Map(nm0000102 -> 3) nm0457663 Map() nm0797616 Map(nm0000102 -> 4) nm0995597 Map() nm0092967 Map(nm0000102 -> 4) nm1576805 Map(nm0000102 -> 4) nm0002347 Map(nm0000102 -> 4) nm0135589 Map(nm0000102 -> 4) nm0706461 Map(nm0000102 -> 4) nm1345123 Map(nm0000102 -> 3) nm1717757 Map(nm0000102 -> 5) nm7414382 Map() nm9025636 Map(nm0000102 -> 4) nm2106007 Map(nm0000102 -> 3) nm1589312 Map(nm0000102 -> 4) nm0344986 Map(nm0000102 -> 4) nm0480754 Map(nm0000102 -> 4) nm0955969 Map(nm0000102 -> 4) nm1715142 Map(nm0000102 -> 5) nm2961569 Map() nm5282143 Map(nm0000102 -> 4) nm0283138 Map(nm0000102 -> 4) nm1507623 Map() nm2900452 Map(nm0000102 -> 3) nm4979616 Map() nm2290876 Map(nm0000102 -> 3) nm0081116 Map(nm0000102 -> 4) nm2423795 Map(nm0000102 -> 6) nm0461109 Map(nm0000102 -> 3) nm4760665 Map(nm0000102 -> 4) nm1634832 Map(nm0000102 -> 2) nm4383483 Map(nm0000102 -> 4) nm0429084 Map() nm0484733 Map(nm0000102 -> 6) nm1090865 Map(nm0000102 -> 3) nm3964972 Map() nm8052326 Map(nm0000102 -> 5) nm0583915 Map(nm0000102 -> 4) nm2410462 Map(nm0000102 -> 5) nm2503061 Map(nm0000102 -> 5) nm3002412 Map(nm0000102 -> 5) nm0077736 Map(nm0000102 -> 5) nm0127960 Map(nm0000

In [0]:
#3.

In [0]:
all_connected = shortest_path.rdd.map(lambda x: (x[0],x[1]['nm0000102']) if len(x[1]) != 0 else (x[0],20))
#this is a paired RDD like (nconst, degree) if the dictionary is not empty we will get the degree value else the value is 20

summary = all_connected.toDF(['Person','Degree'])\
                       .groupBy('Degree')\
                       .count()\
                       .sort('Degree')


My final answear is:

In [0]:
summary.show()

+------+-----+
|Degree|count|
+------+-----+
|     0|    1|
|     1|  133|
|     2| 3547|
|     3|19475|
|     4|31215|
|     5|14930|
|     6| 2782|
|     7|  376|
|     8|   55|
|     9|    9|
|    20|20518|
+------+-----+



### Exploring the data with RDD's

Using RDDs and (not dataframes) answer the following questions (if you loaded your data into spark in a dataframe you can convert to an RDD of rows easily using `.rdd`):

**Q8** Movies can have multiple genres. Considering only titles of the type 'movie' what is the combination of genres that is the most popluar (as measured by number of reviews). Hint: paired RDD's will be useful.

In [0]:
# to answear this question only title_ratings and title_basics are necessary which will be converted to paired RDDs.

In [0]:
RDD_basics = title_basics.rdd.filter(lambda x: (x[1] == 'movie') & (x[8] != '\\N'))\
                             .map(lambda x: (x[0],x[8]))
# we only want the movies and we are not interested in movies without genre

RDD_ratings = title_ratings.rdd.map(lambda x: (x[0],x[2]))

In [0]:
RDD_basics.take(5), RDD_ratings.take(5)

Out[58]: ([('tt0000574', 'Action,Adventure,Biography'),
  ('tt0000591', 'Drama'),
  ('tt0000615', 'Drama'),
  ('tt0000630', 'Drama'),
  ('tt0000675', 'Drama')],
 [('tt0000001', 1884),
  ('tt0000002', 250),
  ('tt0000003', 1669),
  ('tt0000004', 163),
  ('tt0000005', 2490)])

In [0]:
# let's merge both rdds and then transform it into a new paired RDD with (genre,numVotes) and sum the numVotes per genre

In [0]:
RDD_merged = RDD_basics.join(RDD_ratings)
RDD_genre_rating = RDD_merged.map(lambda x: (x[1][0], x[1][1]))\
                             .reduceByKey(lambda a,b: a + b)\
                             .sortBy(lambda x: x[1], ascending = False)

My final answear is:

In [0]:
RDD_genre_rating.take(1)[0]

Out[61]: ('Action,Adventure,Sci-Fi', 51140697)

**Q9** Movies can have multiple genres. Considering only titles of the type 'movie', and movies with more than 400 ratings, what is the combination of genres that has the highest **average movie rating** (you can average the movie rating for each movie in that genre combination). Hint: paired RDD's will be useful.

In [0]:
# pratically we need to do the same as we did above but considering the average rating of the movie instead of the 'numVotes'. In the end, when grouping the ratings by the genre, we will need to calculate the average of the results.

In [0]:
RDD_basics_ = title_basics.rdd.filter(lambda x: (x[1] == 'movie') & (x[8] != '\\N'))\
                              .map(lambda x: (x[0],x[8]))
RDD_ratings_ = title_ratings.rdd.filter(lambda x: (x[2] > 400))\
                            .map(lambda x: (x[0],x[1]))

In [0]:
RDD_merged_ = RDD_basics_.join(RDD_ratings_)
RDD_averages = RDD_merged_.map(lambda x: (x[1][0], x[1][1]))\
                          .groupByKey() \
                          .mapValues(lambda x: sum(x) / len(x)) \   # average the results
                          .sortBy(lambda x: x[1], ascending = False)

My answear is:

In [0]:
RDD_averages.take(1)

Out[65]: [('Music,Musical', 8.4)]

**Q10** Movies can have multiple genres. What is **the individual genre** which is the most popular as meaured by number of votes. Votes for multiple genres count towards each genre listed. Hint: flatmap and pairedRDD's will be useful here.

In [0]:
#using the rdds created in Q8

#flatMapValues: https://spark.apache.org/docs/3.1.1/api/python/reference/api/pyspark.RDD.flatMapValues.html

In [0]:
RDD_individual_genre_rating = RDD_merged.map(lambda x: (x[1][1], x[1][0]))\           # storing it like (numVotes, genres) in order to apply flatMapValues
                                        .flatMapValues(lambda x: x.split(','))\       # split the 'genres' on the ','
                                        .map(lambda x: (x[1],x[0]))\                  # genre as the key to perform reduceByKey
                                        .reduceByKey(lambda a,b: a + b)\
                                        .sortBy(lambda x: x[1], ascending = False)

My answaer is:

In [0]:
RDD_individual_genre_rating.take(1)

Out[68]: [('Drama', 487969283)]

## Engineering the perfect cast
We have created a number of potential features for predicting the rating of a movie based on its cast. Use sparkML to build a simple linear model to predict the rating of a movie based on the following features:

1. The total number of movies in which the actors / actresses have acted (based on Q3)
1. The average pagerank of the cast in each movie (based on Q5)
1. The average outDegree of the cast in each movie (based on Q6)
1. The average value for for the cast of degrees of Kevin Bacon (based on Q7).

You will need to create a dataframe with the required features and label. Use a pipeline to create the vectors required by sparkML and apply the model. Remember to split your dataset, leave 30% of the data for testing, when splitting your data use the option seed=0.

**Q11** Provide the coefficients of the regression and the accuracy of your model on that test dataset according to RSME.

In [0]:
# we need to merge all information in a single dataframe. To do it, we will use 'nconst' and in the end we can drop it.

In [0]:
data_joins =  df.rdd.map(lambda x: (x[0], 1))\
                 .reduceByKey(lambda a,b : a + b)\
                 .toDF(['nconst', 'count'])\          # for 1.
                 .join(df.select(['nconst','tconst']), on = 'nconst')\     # we will need 'tconst' in the future to group
                 .join(pageRank_vertices.select(pageRank_vertices['id'].alias('nconst'),'pagerank'),on = 'nconst')\        # for 3.
                 .join(RDD_outDegrees.toDF(['nconst', 'outDegrees']),on = 'nconst')\       # for 4.
                 .join(all_connected.toDF(['nconst','DegreeKB']),on = 'nconst')\         # for 5.
                 .drop('nconst').cache()

In [0]:
# now that we have all the information needed in a single dataframe we need to group by 'tconst' and aggregate the information as it is asked.

In [0]:
data = data_joins.groupBy('tconst').agg({'count':'sum', 'pagerank':'avg', 'outDegrees':'avg', 'DegreeKB':'avg'})\
                 .join(title_ratings.select(title_ratings['averageRating'].alias('rating'),'tconst'), on = 'tconst')


In [0]:
data.show(10)

+---------+---------------+----------+-------------------+-------------+------+
|   tconst|avg(outDegrees)|sum(count)|      avg(pagerank)|avg(DegreeKB)|rating|
+---------+---------------+----------+-------------------+-------------+------+
|tt0036077|            1.0|        17| 0.1779822304598052|          5.0|   7.5|
|tt0042303|           13.5|        68| 1.8992518523345572|          3.0|   6.6|
|tt0044431|            1.0|        14|0.40176691408112764|          5.0|   6.1|
|tt0044770|            4.0|        20| 1.2466639401145558|          3.5|   6.6|
|tt0045015|            1.0|        12|  0.279047871039105|          4.0|   5.6|
|tt0048006|           13.0|        80|  1.742202628073934|          3.0|   5.9|
|tt0049309|           12.0|        45| 1.7129026129178544|          4.0|   6.0|
|tt0049448|           97.0|        99|   8.35048381562023|          3.0|   6.9|
|tt0050254|           31.0|        62| 3.4561042913959445|          4.0|   6.1|
|tt0050651|            1.0|         7|0.

In [0]:
#checking for possible outliers

In [0]:
CONTINUOUS_FEATURES = data.columns[1:5]
data.select(*CONTINUOUS_FEATURES).summary("mean","stddev","min","1%","5%","50%","95%","99%","max",).show()

+-------+------------------+------------------+------------------+------------------+
|summary|   avg(outDegrees)|        sum(count)|     avg(pagerank)|     avg(DegreeKB)|
+-------+------------------+------------------+------------------+------------------+
|   mean| 34.76085908361231| 68.79224072466913| 2.903557449688704|3.8457588344017646|
| stddev|33.330464459704714|101.97909624315612|2.3952855635422696| 2.602621615294829|
|    min|               1.0|                 2|0.1511677262880853|               0.5|
|     1%|               1.0|                 2|0.2737485961963287|              1.75|
|     5%|               2.0|                 4|0.5634925138378909|               2.0|
|    50%|              25.0|                37| 2.326754637709051|               3.5|
|    95%| 99.33333333333333|               232| 7.022442107032612|               5.0|
|    99%|             146.5|               487|10.875284460286021|              20.0|
|    max|             486.0|              1653| 53.690

The columns "avg(outDegrees)","sum(count)" and "avg(pagerank)" seem to have some outliers.

In [0]:
#checking for missing values to see if we need to drop them or if is better to fill them

#inspiration for this code: https://stackoverflow.com/questions/44413132/count-the-number-of-missing-values-in-a-dataframe-spark

In [0]:
nan_columns = data.select([count(when(isnull(col), col)).alias(col) for col in data.columns]).toPandas()
nan_columns

tconst 
 avg(outDegrees) 
 sum(count) 
 avg(pagerank) 
 avg(DegreeKB) 
 rating 
 
 
 
 
 0 
 0 
 0 
 0 
 0 
 0 
 0

We don't have any missing value.

Pipeline

In [0]:
#Note: i will use MinMaxScaler here(as we did in pratical classes) but as we are using the Linear Regression, really sensible to outliers, we should also tried RobustScaler. I will explore it later on Q14.

In [0]:
vector_assembler = MF.VectorAssembler(  
    inputCols=CONTINUOUS_FEATURES,
    outputCol="vector",
)
 
vector_scaler = MF.MinMaxScaler(  
    inputCol="vector",
    outputCol="vector_scaled",
)

lr =  LinearRegression(
    featuresCol="vector_scaled", labelCol="rating", predictionCol="predictions"
)


movies_pipeline = Pipeline(  
    stages=[vector_assembler, vector_scaler,lr],
)

In [0]:
#train-test split

In [0]:
train, test = data.randomSplit([0.7, 0.3], seed = 0)
train.cache()
test.cache()

Out[74]: DataFrame[tconst: string, avg(outDegrees): double, sum(count): bigint, avg(pagerank): double, avg(DegreeKB): double, rating: double]

In [0]:
ourModel = movies_pipeline.fit(train)
predictions = ourModel.transform(test)

evaluator = RegressionEvaluator(
    labelCol="rating",  
    predictionCol="predictions", 
    metricName="rmse",
)

rmse = evaluator.evaluate(predictions)

My final answear is:

In [0]:
print('RMSE of the model: '+str(rmse)+'\nValue of the intercept of the model: ' +str(ourModel.stages[2].intercept)+ '\nValues of the coefficients of the model: '+str(ourModel.stages[2].coefficients.values))

RMSE of the model: 1.2586001057640366
Value of the intercept of the model: 5.88131831475802
Values of the coefficients of the model: [ 4.62602759 -1.13612923 -6.18224448  0.28724882]


**Q12** What score would your model predict for the 1997 movie Titanic.

In [0]:
titanicID = title_basics.filter((title_basics.primaryTitle == 'Titanic') & (title_basics.startYear == 1997) & (title_basics.titleType == 'movie')).collect()[0][0]
#getting Titanic ID

dfTitanic = data.filter(data.tconst == titanicID).cache()
#dataframe with Titanic information

featTitanic = dfTitanic.drop('rating')
labelTitanic = dfTitanic.select('rating')
# splitting dfTitanic into features, that we will use to test the model in the future, and real label(to compare the results)

predictions_titanic = ourModel.transform(featTitanic)
predictedValue = predictions_titanic.collect()[0]['predictions']
labelValue = labelTitanic.collect()[0]['rating']

My final answer:

In [0]:
print(f" Predicted rating for Titanic {predictedValue} ") #predicted rating
print(f" Rating of Titanic = {labelValue} ") #real rating

 Predicted rating for Titanic 5.807227841387321 
 Rating of Titanic = 7.9 


**Q13** Create dummy variables for each of the top 10 movie genres for Q10. These variable should have a value of 1 if the movie was rated with that genre and 0 otherwise. For example the 1997 movie Titanic should have a 1 in the dummy variable column for Romance, and a 1 in the dummy variable column for Drama, and 0's in all the other dummy variable columns.

Does adding these variable to the regression improve your results? What is the new RMSE and predicted rating for the 1997 movie Titanic.

In [0]:
#creating top10_geres with the RDD already created for indvidual genres and rating

In [0]:
top10_genres = RDD_individual_genre_rating.map(lambda x: x[0]).take(10)
print(top10_genres)

['Drama', 'Action', 'Comedy', 'Adventure', 'Crime', 'Thriller', 'Sci-Fi', 'Romance', 'Mystery', 'Horror']


In [0]:
# to create the dummies variables i inspired myself in the link below.
#https://stackoverflow.com/questions/42805663/e-num-get-dummies-in-pyspark

In [0]:
ourDummies = [when(col("genres").contains(genre),1).otherwise(0).alias(genre) for genre in top10_genres]      
dummiesDF = title_basics.select("tconst", "genres", *ourDummies)

dataDummies = data.join(dummiesDF, on='tconst').cache()

dataDummies.show(10)

+---------+---------------+----------+-------------------+-------------+------+--------------------+-----+------+------+---------+-----+--------+------+-------+-------+------+
|   tconst|avg(outDegrees)|sum(count)|      avg(pagerank)|avg(DegreeKB)|rating|              genres|Drama|Action|Comedy|Adventure|Crime|Thriller|Sci-Fi|Romance|Mystery|Horror|
+---------+---------------+----------+-------------------+-------------+------+--------------------+-----+------+------+---------+-----+--------+------+-------+-------+------+
|tt0036077|            1.0|        17| 0.1779822304598052|          5.0|   7.5|               Drama|    1|     0|     0|        0|    0|       0|     0|      0|      0|     0|
|tt0042303|           13.5|        68| 1.8992518523345572|          3.0|   6.6|               Drama|    1|     0|     0|        0|    0|       0|     0|      0|      0|     0|
|tt0044431|            1.0|        14|0.40176691408112764|          5.0|   6.1|      Comedy,Musical|    0|     0|     1|

In [0]:
print(dataDummies.columns[1:5] + dataDummies.columns[7:])

['avg(outDegrees)', 'sum(count)', 'avg(pagerank)', 'avg(DegreeKB)', 'Drama', 'Action', 'Comedy', 'Adventure', 'Crime', 'Thriller', 'Sci-Fi', 'Romance', 'Mystery', 'Horror']


In [0]:
BINARY_FEATURES = dataDummies.columns[7:] 
CONTINUOUS_FEATURES =dataDummies.columns[1:5]

In [0]:
vector_continuous = MF.VectorAssembler(  
    inputCols=CONTINUOUS_FEATURES,
    outputCol="vector_continuous",
)
 
vector_scaler_minmax = MF.MinMaxScaler(  
    inputCol="vector_continuous",
    outputCol="vector_scaled_minmax",
)

vector_assembler = MF.VectorAssembler(
    inputCols=BINARY_FEATURES 
    + ["vector_scaled_minmax"],
    outputCol="features",
)

lr =  LinearRegression(
    featuresCol="features", labelCol="rating", predictionCol="predictions"
)


movies_pipeline = Pipeline(  
    stages=[vector_continuous,vector_scaler_minmax,vector_assembler,lr],
)

In [0]:
train, test = dataDummies.randomSplit([0.7, 0.3], seed = 0)
train.cache()
test.cache()

ourModel_ = movies_pipeline.fit(train)
predictions_ = ourModel_.transform(test)

dfTitanic_ = dataDummies.filter(dataDummies.tconst == titanicID).cache()

featTitanic = dfTitanic_.drop('rating')
labelTitanic = dfTitanic_.select('rating')

predictions_titanic = ourModel_.transform(featTitanic)
predictedValue = predictions_titanic.collect()[0]['predictions']
labelValue = labelTitanic.collect()[0]['rating']

In [0]:
rmse_ = evaluator.evaluate(predictions_)

My final answer is:

In [0]:
print(f" RMSE of the model {rmse_} ")
print(f" Predicted rating for Titanic {predictedValue} ")
print(f" Rating of Titanic = {labelValue} ")

 RMSE of the model 1.1978463942251345 
 Predicted rating for Titanic 6.145087052276939 
 Rating of Titanic = 7.9 


We can conclude the genres of the movie have meaning to predict the 'average rating' of the movie and in the following question we should keep them (RMSE decreased).

**Q14 - Open Question**: Improve your model by testing different machine learning algorithms, using hyperparameter tuning on these algorithms, changing the included features. What is the RMSE of you final model and what rating does it predict for the 1997 movie Titanic.

We already know that including the top10 genres in our model improved the results and makes sense. We can also think that if a movie has more or less adaptation/'versions', can also have meaning for its rating. Is expected that a movie with more adaptations has a higher rating (because it can mean that movie was popular).

In [0]:
numbVersion = title_akas.select(title_akas['titleId'].alias('tconst')).groupBy('tconst').count()
numbVersion.show(3)
#title_akas -- where we have the information about the adaptations/'versions' of a movie

+---------+-----+
|   tconst|count|
+---------+-----+
|tt0000658|    5|
|tt0000839|    2|
|tt0001170|    2|
+---------+-----+
only showing top 3 rows



In [0]:
data_new = dataDummies.join(numbVersion, on = 'tconst')

BINARY_FEATURES = data_new.columns[7:-1]
CONTINUOUS_FEATURES =data_new.columns[1:5] + [data_new.columns[-1]]     #all previous continuous features and the number of versions

vector_continuous = MF.VectorAssembler(  
    inputCols=CONTINUOUS_FEATURES,
    outputCol="vector_continuous",
)
 
vector_scaler_robust = MF.RobustScaler(  
    inputCol="vector_continuous",
    outputCol="vector_scaled_robust",
)

vector_assembler = MF.VectorAssembler(
    inputCols=BINARY_FEATURES 
    + ["vector_scaled_robust"],
    outputCol="features",
)

#checking also the results with the RobustScaler for linear regression which might be more appropriate in our case
#note: parameters of Gradient-boosted tree regressor explained below

lr =  LinearRegression(
    featuresCol="features", labelCol="rating", predictionCol="predictions"
)

rf = RandomForestRegressor(
    featuresCol="features", labelCol="rating", predictionCol="predictions"
)

gbt = GBTRegressor(
    featuresCol="features", labelCol="rating", predictionCol="predictions",maxDepth=5,maxIter = 10
)

movies_pipeline_1 = Pipeline(  
    stages=[vector_continuous,vector_scaler_robust,vector_assembler,lr],
)

movies_pipeline_2 = Pipeline(  
    stages=[vector_continuous,vector_scaler_robust,vector_assembler,rf],
)

movies_pipeline_3 = Pipeline(  
    stages=[vector_continuous,vector_scaler_robust,vector_assembler, gbt],
)

evaluator = RegressionEvaluator(
    labelCol="rating",  
    predictionCol="predictions", 
    metricName="rmse",
)

In [0]:
train, test = data_new.randomSplit([0.7, 0.3], seed = 0)
train.cache()
test.cache()

dfTitanic_NEW = data_new.filter(data_new.tconst == titanicID).cache()

featTitanic = dfTitanic_NEW.drop('rating')
labelTitanic = dfTitanic_NEW.select('rating')

For Linear Model

In [0]:
model_lr = movies_pipeline_1.fit(train)
predictions1 = model_lr.transform(test)
rmse1 = evaluator.evaluate(predictions1)
predictions_Titanic1 = model_lr.transform(featTitanic)

print('RMSE of the model: ' +str(rmse1)+ '\nPredicted rating: ' +str(predictions_Titanic1.collect()[0]['predictions']))

RMSE of the model: 1.1778759514052892
Predicted rating: 7.380385996092974


For Random Forest

In [0]:
model_rf = movies_pipeline_2.fit(train)
predictions2 = model_rf.transform(test)
rmse2 = evaluator.evaluate(predictions2)

In [0]:
predictions_Titanic2 = model_rf.transform(featTitanic)

print('RMSE of the model: ' +str(rmse2)+ '\nPredicted rating: ' +str(predictions_Titanic2.collect()[0]['predictions']))

RMSE of the model: 1.1796799439607009
Predicted rating: 6.635893289095067


For Gradient-boosted tree

This model was really difficult to run, with the cluster dying after a while, I had to reduce my parameters and do some experiments, but I took an interesting conclusion: \
 -first I tried to run with the deafult parameters, was impossible;\
 -then I even tried on a sample of my dataframe but the results were poor becuase the size was too small to can run it;\
 -so, I tried to change the default parameters mainly max depth to 2 and max iteration only 5, now I could run it but not good results;\
 -finally, as my results were improving I changed the max depth to the default value (5) and increase the max iterations to 10, and the results improved being my best model, so if I could further explore the parameteres, probably I would get better results.

In [0]:
model_gbt = movies_pipeline_3.fit(train)

In [0]:
predictions_Titanic3 = model_gbt.transform(featTitanic)

In [0]:
predictions3 = model_gbt.transform(test)
rmse3 = evaluator.evaluate(predictions3)

In [0]:
predictions_Titanic3 = model_gbt.transform(featTitanic)

print('RMSE of the model: ' +str(rmse3)+ '\nPredicted rating: ' +str(predictions_Titanic3.collect()[0]['predictions']))

RMSE of the model: 1.1600398440373452
Predicted rating: 7.467804159050522


Grid Search CV

To do this exercise properly I should run a GridSearch for the three models choosen and compare the resuls. However, cluster dies running a GridSearch with CV for the ensemblers chosen (for Gradient-boosted tree is impossible for me with the community version and for randomm forest i had to limit the parameters to a really small number). So, i will only run a GridSearchCV for Linear Regression to try to improve more the results (this model is really fast comparing to the others).

In [0]:
grid_search = (
    ParamGridBuilder()\
#    .addGrid(lr.regParam, [0.1, 0.01]) \
    .addGrid(lr.elasticNetParam, [0.0, 0.5, 1.0])\
    .build()
)

#grid search for RF
# grid_search = (
#     ParamGridBuilder() 
#     .addGrid(rf.numTrees, [25,50])
#     .addGrid(rf.maxDepth, [2, 4, 10])
#     .build()
# )

#grid search for gbt
# grid_search = (
#     ParamGridBuilder() 
#     .addGrid(gbt.maxIter, [10, 20, 30])
#     .addGrid(gbt.maxDepth, [5, 10])
#     .build()
# )


cv = CrossValidator(
    estimator=movies_pipeline_1,
    estimatorParamMaps=grid_search,
    evaluator=evaluator,
    numFolds=2,
    seed=13,
)

In [0]:
cv_model = cv.fit(train)


In [0]:
predicts = cv_model.transform(test)

rmse1_update = evaluator.evaluate(predicts)
predicts_Titanic = cv_model.transform(featTitanic)

print('RMSE of the model: ' +str(rmse1_update)+ '\nPredicted rating: ' +str(predicts_Titanic.collect()[0]['predictions']))

RMSE of the model: 1.1778759514052894
Predicted rating: 7.380385996092976


Final Results

In [0]:
simpleDF = spark.createDataFrame(
    [
        (1, "lr + GridSearchCV", rmse1_update, predicts_Titanic.collect()[0]['predictions'],labelTitanic.collect()[0]['rating']),  
        (2, "rf", rmse2, predictions_Titanic2.collect()[0]['predictions'],labelTitanic.collect()[0]['rating']),
        (3, "gbt", rmse3, predictions_Titanic3.collect()[0]['predictions'],labelTitanic.collect()[0]['rating']),
    ],
    ["id", "model","RMSE","predTitanic","ratingTtanic"] 
)

simpleDF.show()

+---+-----------------+------------------+-----------------+------------+
| id|            model|              RMSE|      predTitanic|ratingTtanic|
+---+-----------------+------------------+-----------------+------------+
|  1|lr + GridSearchCV|1.1778759514052894|7.380385996092976|         7.9|
|  2|               rf|1.1796799439607009|6.635893289095067|         7.9|
|  3|              gbt|1.1600398440373452|7.467804159050522|         7.9|
+---+-----------------+------------------+-----------------+------------+



Final Conclusions:

If I have to pick one model, I would choose the Gradient-boosted tree regressor, it achieved good results (RMSE = 1.16 and predicted rating for Titanic of 7.47). For this model, I have to say that I really would like to see how far it could go if I could run it properly because the results obtained were increasing as I increased the parameters but I was limitated by the time and the community version.
Random Forest Regressor was indeed, the worst model however, with Grid Search CV we could probably improve the results. 
Finally, I would like to highlight my final Linear Regression with Robust Scaler which also achieve good results and the improvement from Q13 is visible. We can even see that we had no improvement using the GridSearchCV, which it's not strange at all since the search space was very limited.

One final note, the feature included was also important to achieve these results because I made some tests with it and without and the results were favourable. So, as expected, the number of versions/adaptations of a movie has meaning to predict the average rating of a movie.